// @gnss/rtcm/build/messages/observations/glonass.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmMessageGlonassL1L2ObservationsExtended = exports.RtcmMessageGlonassL1L2Observations = exports.RtcmMessageGlonassL1ObservationsExtended = exports.RtcmMessageGlonassL1Observations = exports.GlonassL1L2ObservationsExtendedSatelliteData = exports.GlonassL1L2ObservationsSatelliteData = exports.GlonassL1ObservationsExtendedSatelliteData = exports.GlonassL1ObservationsSatelliteData = exports.GlonassObservationsSatelliteData = exports.GlonassL2ObservationsExtendedData = exports.GlonassL2ObservationsData = exports.GlonassL1ObservationsExtendedData = exports.GlonassL1ObservationsData = exports.GlonassObservationsData = exports.RtcmMessageGlonassObservations = exports.GlonassL2CodeIndicator = exports.GlonassL1CodeIndicator = void 0;
const decode_encode_1 = require("../../decode-encode");
const rtcm_1 = require("../../rtcm");
const utils_1 = require("../../utils");
const common_1 = require("./common");
var GlonassL1CodeIndicator;
(function (GlonassL1CodeIndicator) {
    GlonassL1CodeIndicator[GlonassL1CodeIndicator["CA_CODE"] = 0] = "CA_CODE";
    GlonassL1CodeIndicator[GlonassL1CodeIndicator["P_Y_CODE_DIRECT"] = 1] = "P_Y_CODE_DIRECT";
})(GlonassL1CodeIndicator = exports.GlonassL1CodeIndicator || (exports.GlonassL1CodeIndicator = {}));
var GlonassL2CodeIndicator;
(function (GlonassL2CodeIndicator) {
    GlonassL2CodeIndicator[GlonassL2CodeIndicator["CA_L2C_CODE"] = 0] = "CA_L2C_CODE";
    GlonassL2CodeIndicator[GlonassL2CodeIndicator["P_Y_CODE_DIRECT"] = 1] = "P_Y_CODE_DIRECT";
    GlonassL2CodeIndicator[GlonassL2CodeIndicator["RESERVED_0"] = 2] = "RESERVED_0";
    GlonassL2CodeIndicator[GlonassL2CodeIndicator["RESERVED_1"] = 3] = "RESERVED_1";
})(GlonassL2CodeIndicator = exports.GlonassL2CodeIndicator || (exports.GlonassL2CodeIndicator = {}));
let RtcmMessageGlonassObservations = class RtcmMessageGlonassObservations extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(3),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGlonassObservations.prototype, "referenceStationId", void 0);
__decorate([
    decode_encode_1.DF(34),
    decode_encode_1.UInt(27),
    __metadata("design:type", Number)
], RtcmMessageGlonassObservations.prototype, "glonassEpochTime", void 0);
__decorate([
    decode_encode_1.DF(5),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageGlonassObservations.prototype, "synchronousGnss", void 0);
__decorate([
    decode_encode_1.DF(35),
    decode_encode_1.ArrLength(5, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageGlonassObservations.prototype, "_satelliteCount", void 0);
__decorate([
    decode_encode_1.DF(36),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageGlonassObservations.prototype, "divergenceFreeSmoothing", void 0);
__decorate([
    decode_encode_1.DF(37),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], RtcmMessageGlonassObservations.prototype, "smoothingInterval", void 0);
RtcmMessageGlonassObservations = __decorate([
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_0),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageGlonassObservations);
exports.RtcmMessageGlonassObservations = RtcmMessageGlonassObservations;
class GlonassObservationsData {
}
exports.GlonassObservationsData = GlonassObservationsData;
GlonassObservationsData.construct = utils_1.constructWithPropertiesPlus;
class GlonassL1ObservationsData extends GlonassObservationsData {
    get lockTimeMillis() { return common_1.legacyLockIndicatorToTime(this.lockTimeIndicator); }
    set lockTimeMillis(time) { this.lockTimeIndicator = common_1.legacyLockTimeToIndicator(time); }
}
__decorate([
    decode_encode_1.DF(39),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], GlonassL1ObservationsData.prototype, "codeIndicator", void 0);
__decorate([
    decode_encode_1.DF(40),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], GlonassL1ObservationsData.prototype, "satelliteFrequencyChannelNumber", void 0);
__decorate([
    decode_encode_1.DF(41),
    decode_encode_1.UInt(25),
    __metadata("design:type", Number)
], GlonassL1ObservationsData.prototype, "pseudorange", void 0);
__decorate([
    decode_encode_1.DF(42),
    decode_encode_1.Int(20),
    __metadata("design:type", Number)
], GlonassL1ObservationsData.prototype, "phaserangePseudorangeDiff", void 0);
__decorate([
    decode_encode_1.DF(43),
    decode_encode_1.UInt(7),
    __metadata("design:type", Number)
], GlonassL1ObservationsData.prototype, "lockTimeIndicator", void 0);
exports.GlonassL1ObservationsData = GlonassL1ObservationsData;
class GlonassL1ObservationsExtendedData extends GlonassL1ObservationsData {
}
__decorate([
    decode_encode_1.DF(44),
    decode_encode_1.UInt(7),
    __metadata("design:type", Number)
], GlonassL1ObservationsExtendedData.prototype, "pseudorangeModulusAmbiguity", void 0);
__decorate([
    decode_encode_1.DF(45),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GlonassL1ObservationsExtendedData.prototype, "cnr", void 0);
exports.GlonassL1ObservationsExtendedData = GlonassL1ObservationsExtendedData;
class GlonassL2ObservationsData extends GlonassObservationsData {
    get lockTimeMillis() { return common_1.legacyLockIndicatorToTime(this.lockTimeIndicator); }
    set lockTimeMillis(time) { this.lockTimeIndicator = common_1.legacyLockTimeToIndicator(time); }
}
__decorate([
    decode_encode_1.DF(46),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], GlonassL2ObservationsData.prototype, "codeIndicator", void 0);
__decorate([
    decode_encode_1.DF(47),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], GlonassL2ObservationsData.prototype, "pseudorangeDiff", void 0);
__decorate([
    decode_encode_1.DF(48),
    decode_encode_1.Int(20),
    __metadata("design:type", Number)
], GlonassL2ObservationsData.prototype, "phaserangePseudorangeDiff", void 0);
__decorate([
    decode_encode_1.DF(49),
    decode_encode_1.UInt(7),
    __metadata("design:type", Number)
], GlonassL2ObservationsData.prototype, "lockTimeIndicator", void 0);
exports.GlonassL2ObservationsData = GlonassL2ObservationsData;
class GlonassL2ObservationsExtendedData extends GlonassL2ObservationsData {
}
__decorate([
    decode_encode_1.DF(50),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GlonassL2ObservationsExtendedData.prototype, "cnr", void 0);
exports.GlonassL2ObservationsExtendedData = GlonassL2ObservationsExtendedData;
class GlonassObservationsSatelliteData {
}
GlonassObservationsSatelliteData.construct = utils_1.constructWithProperties;
__decorate([
    decode_encode_1.DF(38),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GlonassObservationsSatelliteData.prototype, "satelliteId", void 0);
exports.GlonassObservationsSatelliteData = GlonassObservationsSatelliteData;
class GlonassL1ObservationsSatelliteData extends GlonassObservationsSatelliteData {
}
__decorate([
    decode_encode_1.Obj(GlonassL1ObservationsData),
    __metadata("design:type", GlonassL1ObservationsData)
], GlonassL1ObservationsSatelliteData.prototype, "l1", void 0);
exports.GlonassL1ObservationsSatelliteData = GlonassL1ObservationsSatelliteData;
class GlonassL1ObservationsExtendedSatelliteData extends GlonassObservationsSatelliteData {
}
__decorate([
    decode_encode_1.Obj(GlonassL1ObservationsExtendedData),
    __metadata("design:type", GlonassL1ObservationsExtendedData)
], GlonassL1ObservationsExtendedSatelliteData.prototype, "l1", void 0);
exports.GlonassL1ObservationsExtendedSatelliteData = GlonassL1ObservationsExtendedSatelliteData;
class GlonassL1L2ObservationsSatelliteData extends GlonassObservationsSatelliteData {
    toL1Only() {
        return GlonassL1ObservationsSatelliteData.construct({ satelliteId: this.satelliteId, l1: this.l1 });
    }
}
__decorate([
    decode_encode_1.Obj(GlonassL1ObservationsData),
    __metadata("design:type", GlonassL1ObservationsData)
], GlonassL1L2ObservationsSatelliteData.prototype, "l1", void 0);
__decorate([
    decode_encode_1.Obj(GlonassL2ObservationsData),
    __metadata("design:type", GlonassL2ObservationsData)
], GlonassL1L2ObservationsSatelliteData.prototype, "l2", void 0);
exports.GlonassL1L2ObservationsSatelliteData = GlonassL1L2ObservationsSatelliteData;
class GlonassL1L2ObservationsExtendedSatelliteData extends GlonassObservationsSatelliteData {
    toL1Only() {
        return GlonassL1ObservationsExtendedSatelliteData.construct({ satelliteId: this.satelliteId, l1: this.l1 });
    }
}
__decorate([
    decode_encode_1.Obj(GlonassL1ObservationsExtendedData),
    __metadata("design:type", GlonassL1ObservationsExtendedData)
], GlonassL1L2ObservationsExtendedSatelliteData.prototype, "l1", void 0);
__decorate([
    decode_encode_1.Obj(GlonassL2ObservationsExtendedData),
    __metadata("design:type", GlonassL2ObservationsExtendedData)
], GlonassL1L2ObservationsExtendedSatelliteData.prototype, "l2", void 0);
exports.GlonassL1L2ObservationsExtendedSatelliteData = GlonassL1L2ObservationsExtendedSatelliteData;
let RtcmMessageGlonassL1Observations = class RtcmMessageGlonassL1Observations extends RtcmMessageGlonassObservations {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassL1ObservationsSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGlonassL1Observations.prototype, "satellites", void 0);
RtcmMessageGlonassL1Observations = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_L1_OBSERVATIONS)
], RtcmMessageGlonassL1Observations);
exports.RtcmMessageGlonassL1Observations = RtcmMessageGlonassL1Observations;
let RtcmMessageGlonassL1ObservationsExtended = class RtcmMessageGlonassL1ObservationsExtended extends RtcmMessageGlonassObservations {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassL1ObservationsExtendedSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGlonassL1ObservationsExtended.prototype, "satellites", void 0);
RtcmMessageGlonassL1ObservationsExtended = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_L1_OBSERVATIONS_EXTENDED)
], RtcmMessageGlonassL1ObservationsExtended);
exports.RtcmMessageGlonassL1ObservationsExtended = RtcmMessageGlonassL1ObservationsExtended;
let RtcmMessageGlonassL1L2Observations = class RtcmMessageGlonassL1L2Observations extends RtcmMessageGlonassObservations {
    toL1Only() {
        return RtcmMessageGlonassL1Observations.construct(Object.assign(Object.assign({}, this), { satellites: this.satellites.map(s => s.toL1Only()) }));
    }
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassL1L2ObservationsSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGlonassL1L2Observations.prototype, "satellites", void 0);
RtcmMessageGlonassL1L2Observations = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_L1_L2_OBSERVATIONS)
], RtcmMessageGlonassL1L2Observations);
exports.RtcmMessageGlonassL1L2Observations = RtcmMessageGlonassL1L2Observations;
let RtcmMessageGlonassL1L2ObservationsExtended = class RtcmMessageGlonassL1L2ObservationsExtended extends RtcmMessageGlonassObservations {
    toL1Only() {
        return RtcmMessageGlonassL1ObservationsExtended.construct(Object.assign(Object.assign({}, this), { satellites: this.satellites.map(s => s.toL1Only()) }));
    }
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassL1L2ObservationsExtendedSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGlonassL1L2ObservationsExtended.prototype, "satellites", void 0);
RtcmMessageGlonassL1L2ObservationsExtended = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_L1_L2_OBSERVATIONS_EXTENDED)
], RtcmMessageGlonassL1L2ObservationsExtended);
exports.RtcmMessageGlonassL1L2ObservationsExtended = RtcmMessageGlonassL1L2ObservationsExtended;
