// @gnss/rtcm/build/messages/observations/gps.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmMessageGpsL1L2ObservationsExtended = exports.RtcmMessageGpsL1L2Observations = exports.RtcmMessageGpsL1ObservationsExtended = exports.RtcmMessageGpsL1Observations = exports.GpsL1L2ObservationsExtendedSatelliteData = exports.GpsL1L2ObservationsSatelliteData = exports.GpsL1ObservationsExtendedSatelliteData = exports.GpsL1ObservationsSatelliteData = exports.GpsObservationsSatelliteData = exports.GpsL2ObservationsExtendedData = exports.GpsL2ObservationsData = exports.GpsL1ObservationsExtendedData = exports.GpsL1ObservationsData = exports.GpsObservationsData = exports.RtcmMessageGpsObservations = exports.GpsL2CodeIndicator = exports.GpsL1CodeIndicator = void 0;
const decode_encode_1 = require("../../decode-encode");
const rtcm_1 = require("../../rtcm");
const constants_1 = require("../../constants");
const common_1 = require("./common");
const utils_1 = require("../../utils");
var GpsL1CodeIndicator;
(function (GpsL1CodeIndicator) {
    GpsL1CodeIndicator[GpsL1CodeIndicator["CA_CODE"] = 0] = "CA_CODE";
    GpsL1CodeIndicator[GpsL1CodeIndicator["P_Y_CODE_DIRECT"] = 1] = "P_Y_CODE_DIRECT";
})(GpsL1CodeIndicator = exports.GpsL1CodeIndicator || (exports.GpsL1CodeIndicator = {}));
var GpsL2CodeIndicator;
(function (GpsL2CodeIndicator) {
    GpsL2CodeIndicator[GpsL2CodeIndicator["CA_L2C_CODE"] = 0] = "CA_L2C_CODE";
    GpsL2CodeIndicator[GpsL2CodeIndicator["P_Y_CODE_DIRECT"] = 1] = "P_Y_CODE_DIRECT";
    GpsL2CodeIndicator[GpsL2CodeIndicator["P_Y_CODE_CROSS_CORRELATED"] = 2] = "P_Y_CODE_CROSS_CORRELATED";
    GpsL2CodeIndicator[GpsL2CodeIndicator["CORRELATED_P_Y"] = 3] = "CORRELATED_P_Y";
})(GpsL2CodeIndicator = exports.GpsL2CodeIndicator || (exports.GpsL2CodeIndicator = {}));
let RtcmMessageGpsObservations = class RtcmMessageGpsObservations extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(3),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGpsObservations.prototype, "referenceStationId", void 0);
__decorate([
    decode_encode_1.DF(4),
    decode_encode_1.UInt(30),
    __metadata("design:type", Number)
], RtcmMessageGpsObservations.prototype, "gpsEpochTime", void 0);
__decorate([
    decode_encode_1.DF(5),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageGpsObservations.prototype, "synchronousGnss", void 0);
__decorate([
    decode_encode_1.DF(6),
    decode_encode_1.ArrLength(5, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageGpsObservations.prototype, "_satelliteCount", void 0);
__decorate([
    decode_encode_1.DF(7),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageGpsObservations.prototype, "divergenceFreeSmoothing", void 0);
__decorate([
    decode_encode_1.DF(8),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], RtcmMessageGpsObservations.prototype, "smoothingInterval", void 0);
RtcmMessageGpsObservations = __decorate([
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageGpsObservations);
exports.RtcmMessageGpsObservations = RtcmMessageGpsObservations;
class GpsObservationsData {
}
exports.GpsObservationsData = GpsObservationsData;
GpsObservationsData.construct = utils_1.constructWithPropertiesPlus;
class GpsL1ObservationsData extends GpsObservationsData {
    get lockTimeMillis() { return common_1.legacyLockIndicatorToTime(this.lockTimeIndicator); }
    set lockTimeMillis(time) { this.lockTimeIndicator = common_1.legacyLockTimeToIndicator(time); }
}
__decorate([
    decode_encode_1.DF(10),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], GpsL1ObservationsData.prototype, "codeIndicator", void 0);
__decorate([
    decode_encode_1.DF(11),
    decode_encode_1.UInt(24),
    __metadata("design:type", Number)
], GpsL1ObservationsData.prototype, "pseudorange", void 0);
__decorate([
    decode_encode_1.DF(12),
    decode_encode_1.Int(20),
    __metadata("design:type", Number)
], GpsL1ObservationsData.prototype, "phaserangePseudorangeDiff", void 0);
__decorate([
    decode_encode_1.DF(13),
    decode_encode_1.UInt(7),
    __metadata("design:type", Number)
], GpsL1ObservationsData.prototype, "lockTimeIndicator", void 0);
exports.GpsL1ObservationsData = GpsL1ObservationsData;
class GpsL1ObservationsExtendedData extends GpsL1ObservationsData {
    get totalPseudorangeMeters() { return this.pseudorangeModulusAmbiguity * constants_1.SPEED_OF_LIGHT / 1000 + this.pseudorange / 0.02; }
    set totalPseudorangeMeters(pseudorange) {
        this.pseudorange = (pseudorange % (constants_1.SPEED_OF_LIGHT / 1000)) * 0.02;
        this.pseudorangeModulusAmbiguity = Math.floor(pseudorange * 1000 / constants_1.SPEED_OF_LIGHT);
    }
}
__decorate([
    decode_encode_1.DF(14),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GpsL1ObservationsExtendedData.prototype, "pseudorangeModulusAmbiguity", void 0);
__decorate([
    decode_encode_1.DF(15),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GpsL1ObservationsExtendedData.prototype, "cnr", void 0);
exports.GpsL1ObservationsExtendedData = GpsL1ObservationsExtendedData;
class GpsL2ObservationsData extends GpsObservationsData {
    get lockTimeMillis() { return common_1.legacyLockIndicatorToTime(this.lockTimeIndicator); }
    set lockTimeMillis(time) { this.lockTimeIndicator = common_1.legacyLockTimeToIndicator(time); }
}
__decorate([
    decode_encode_1.DF(16),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], GpsL2ObservationsData.prototype, "codeIndicator", void 0);
__decorate([
    decode_encode_1.DF(17),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], GpsL2ObservationsData.prototype, "pseudorangeDiff", void 0);
__decorate([
    decode_encode_1.DF(18),
    decode_encode_1.Int(20),
    __metadata("design:type", Number)
], GpsL2ObservationsData.prototype, "phaserangePseudorangeDiff", void 0);
__decorate([
    decode_encode_1.DF(19),
    decode_encode_1.UInt(7),
    __metadata("design:type", Number)
], GpsL2ObservationsData.prototype, "lockTimeIndicator", void 0);
exports.GpsL2ObservationsData = GpsL2ObservationsData;
class GpsL2ObservationsExtendedData extends GpsL2ObservationsData {
}
__decorate([
    decode_encode_1.DF(20),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GpsL2ObservationsExtendedData.prototype, "cnr", void 0);
exports.GpsL2ObservationsExtendedData = GpsL2ObservationsExtendedData;
class GpsObservationsSatelliteData {
}
GpsObservationsSatelliteData.construct = utils_1.constructWithProperties;
__decorate([
    decode_encode_1.DF(9),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsObservationsSatelliteData.prototype, "satelliteId", void 0);
exports.GpsObservationsSatelliteData = GpsObservationsSatelliteData;
class GpsL1ObservationsSatelliteData extends GpsObservationsSatelliteData {
}
__decorate([
    decode_encode_1.Obj(GpsL1ObservationsData),
    __metadata("design:type", GpsL1ObservationsData)
], GpsL1ObservationsSatelliteData.prototype, "l1", void 0);
exports.GpsL1ObservationsSatelliteData = GpsL1ObservationsSatelliteData;
class GpsL1ObservationsExtendedSatelliteData extends GpsObservationsSatelliteData {
}
__decorate([
    decode_encode_1.Obj(GpsL1ObservationsExtendedData),
    __metadata("design:type", GpsL1ObservationsExtendedData)
], GpsL1ObservationsExtendedSatelliteData.prototype, "l1", void 0);
exports.GpsL1ObservationsExtendedSatelliteData = GpsL1ObservationsExtendedSatelliteData;
class GpsL1L2ObservationsSatelliteData extends GpsObservationsSatelliteData {
    toL1Only() {
        return GpsL1ObservationsSatelliteData.construct({ satelliteId: this.satelliteId, l1: this.l1 });
    }
}
__decorate([
    decode_encode_1.Obj(GpsL1ObservationsData),
    __metadata("design:type", GpsL1ObservationsData)
], GpsL1L2ObservationsSatelliteData.prototype, "l1", void 0);
__decorate([
    decode_encode_1.Obj(GpsL2ObservationsData),
    __metadata("design:type", GpsL2ObservationsData)
], GpsL1L2ObservationsSatelliteData.prototype, "l2", void 0);
exports.GpsL1L2ObservationsSatelliteData = GpsL1L2ObservationsSatelliteData;
class GpsL1L2ObservationsExtendedSatelliteData extends GpsObservationsSatelliteData {
    toL1Only() {
        return GpsL1ObservationsExtendedSatelliteData.construct({ satelliteId: this.satelliteId, l1: this.l1 });
    }
}
__decorate([
    decode_encode_1.Obj(GpsL1ObservationsExtendedData),
    __metadata("design:type", GpsL1ObservationsExtendedData)
], GpsL1L2ObservationsExtendedSatelliteData.prototype, "l1", void 0);
__decorate([
    decode_encode_1.Obj(GpsL2ObservationsExtendedData),
    __metadata("design:type", GpsL2ObservationsExtendedData)
], GpsL1L2ObservationsExtendedSatelliteData.prototype, "l2", void 0);
exports.GpsL1L2ObservationsExtendedSatelliteData = GpsL1L2ObservationsExtendedSatelliteData;
let RtcmMessageGpsL1Observations = class RtcmMessageGpsL1Observations extends RtcmMessageGpsObservations {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsL1ObservationsSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGpsL1Observations.prototype, "satellites", void 0);
RtcmMessageGpsL1Observations = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_L1_OBSERVATIONS)
], RtcmMessageGpsL1Observations);
exports.RtcmMessageGpsL1Observations = RtcmMessageGpsL1Observations;
let RtcmMessageGpsL1ObservationsExtended = class RtcmMessageGpsL1ObservationsExtended extends RtcmMessageGpsObservations {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsL1ObservationsExtendedSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGpsL1ObservationsExtended.prototype, "satellites", void 0);
RtcmMessageGpsL1ObservationsExtended = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_L1_OBSERVATIONS_EXTENDED)
], RtcmMessageGpsL1ObservationsExtended);
exports.RtcmMessageGpsL1ObservationsExtended = RtcmMessageGpsL1ObservationsExtended;
let RtcmMessageGpsL1L2Observations = class RtcmMessageGpsL1L2Observations extends RtcmMessageGpsObservations {
    toL1Only() {
        return RtcmMessageGpsL1Observations.construct(Object.assign(Object.assign({}, this), { satellites: this.satellites.map(s => s.toL1Only()) }));
    }
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsL1L2ObservationsSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGpsL1L2Observations.prototype, "satellites", void 0);
RtcmMessageGpsL1L2Observations = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_L1_L2_OBSERVATIONS)
], RtcmMessageGpsL1L2Observations);
exports.RtcmMessageGpsL1L2Observations = RtcmMessageGpsL1L2Observations;
let RtcmMessageGpsL1L2ObservationsExtended = class RtcmMessageGpsL1L2ObservationsExtended extends RtcmMessageGpsObservations {
    toL1Only() {
        return RtcmMessageGpsL1ObservationsExtended.construct(Object.assign(Object.assign({}, this), { satellites: this.satellites.map(s => s.toL1Only()) }));
    }
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsL1L2ObservationsExtendedSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGpsL1L2ObservationsExtended.prototype, "satellites", void 0);
RtcmMessageGpsL1L2ObservationsExtended = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_L1_L2_OBSERVATIONS_EXTENDED)
], RtcmMessageGpsL1L2ObservationsExtended);
exports.RtcmMessageGpsL1L2ObservationsExtended = RtcmMessageGpsL1L2ObservationsExtended;
