// @gnss/rtcm/build/messages/network/glonass.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmMessageGlonassNetworkFkpGradient = exports.GlonassNetworkFkpGradientSatelliteData = exports.RtcmMessageGlonassNetworkRtkResidual = exports.GlonassNetworkRtkResidualSatelliteData = exports.RtcmMessageGlonassCombinedCorrectionDifferences = exports.RtcmMessageGlonassGeometricCorrectionDifferences = exports.RtcmMessageGlonassIonosphericCorrectionDifferences = exports.GlonassCombinedCorrectionDifferencesSatelliteData = exports.GlonassGeometricCorrectionDifferencesSatelliteData = exports.GlonassIonosphericCorrectionDifferencesSatelliteData = exports.GlonassCorrectionDifferencesBaseSatelliteData = exports.RtcmMessageGlonassCorrectionDifferences = void 0;
const decode_encode_1 = require("../../decode-encode");
const rtcm_1 = require("../../rtcm");
let RtcmMessageGlonassCorrectionDifferences = class RtcmMessageGlonassCorrectionDifferences extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(59),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], RtcmMessageGlonassCorrectionDifferences.prototype, "networkId", void 0);
__decorate([
    decode_encode_1.DF(72),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageGlonassCorrectionDifferences.prototype, "subnetworkId", void 0);
__decorate([
    decode_encode_1.DF(233),
    decode_encode_1.UInt(20),
    __metadata("design:type", Number)
], RtcmMessageGlonassCorrectionDifferences.prototype, "glonassEpochTime", void 0);
__decorate([
    decode_encode_1.DF(66),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageGlonassCorrectionDifferences.prototype, "multipleMessageIndicator", void 0);
__decorate([
    decode_encode_1.DF(60),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGlonassCorrectionDifferences.prototype, "masterReferenceStationId", void 0);
__decorate([
    decode_encode_1.DF(61),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGlonassCorrectionDifferences.prototype, "auxiliaryReferenceStationId", void 0);
__decorate([
    decode_encode_1.DF(234),
    decode_encode_1.ArrLength(4, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageGlonassCorrectionDifferences.prototype, "_satelliteCount", void 0);
RtcmMessageGlonassCorrectionDifferences = __decorate([
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageGlonassCorrectionDifferences);
exports.RtcmMessageGlonassCorrectionDifferences = RtcmMessageGlonassCorrectionDifferences;
class GlonassCorrectionDifferencesBaseSatelliteData {
}
__decorate([
    decode_encode_1.DF(38),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GlonassCorrectionDifferencesBaseSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(235),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], GlonassCorrectionDifferencesBaseSatelliteData.prototype, "ambiguityStatusFlag", void 0);
__decorate([
    decode_encode_1.DF(236),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], GlonassCorrectionDifferencesBaseSatelliteData.prototype, "nonSyncCount", void 0);
exports.GlonassCorrectionDifferencesBaseSatelliteData = GlonassCorrectionDifferencesBaseSatelliteData;
class GlonassIonosphericCorrectionDifferencesSatelliteData extends GlonassCorrectionDifferencesBaseSatelliteData {
}
__decorate([
    decode_encode_1.DF(237),
    decode_encode_1.Int(17),
    __metadata("design:type", Number)
], GlonassIonosphericCorrectionDifferencesSatelliteData.prototype, "ionosphericCarrierPhaseCorrectionDifference", void 0);
exports.GlonassIonosphericCorrectionDifferencesSatelliteData = GlonassIonosphericCorrectionDifferencesSatelliteData;
class GlonassGeometricCorrectionDifferencesSatelliteData extends GlonassCorrectionDifferencesBaseSatelliteData {
}
__decorate([
    decode_encode_1.DF(238),
    decode_encode_1.Int(17),
    __metadata("design:type", Number)
], GlonassGeometricCorrectionDifferencesSatelliteData.prototype, "geometricCarrierPhaseCorrectionDifference", void 0);
__decorate([
    decode_encode_1.DF(239),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GlonassGeometricCorrectionDifferencesSatelliteData.prototype, "iod", void 0);
exports.GlonassGeometricCorrectionDifferencesSatelliteData = GlonassGeometricCorrectionDifferencesSatelliteData;
class GlonassCombinedCorrectionDifferencesSatelliteData extends GlonassGeometricCorrectionDifferencesSatelliteData {
}
__decorate([
    decode_encode_1.DF(237),
    decode_encode_1.Int(17),
    __metadata("design:type", Number)
], GlonassCombinedCorrectionDifferencesSatelliteData.prototype, "ionosphericCarrierPhaseCorrectionDifference", void 0);
exports.GlonassCombinedCorrectionDifferencesSatelliteData = GlonassCombinedCorrectionDifferencesSatelliteData;
let RtcmMessageGlonassIonosphericCorrectionDifferences = class RtcmMessageGlonassIonosphericCorrectionDifferences extends RtcmMessageGlonassCorrectionDifferences {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassIonosphericCorrectionDifferencesSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGlonassIonosphericCorrectionDifferences.prototype, "satellites", void 0);
RtcmMessageGlonassIonosphericCorrectionDifferences = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_IONOSPHERIC_CORRECTION_DIFFERENCES)
], RtcmMessageGlonassIonosphericCorrectionDifferences);
exports.RtcmMessageGlonassIonosphericCorrectionDifferences = RtcmMessageGlonassIonosphericCorrectionDifferences;
let RtcmMessageGlonassGeometricCorrectionDifferences = class RtcmMessageGlonassGeometricCorrectionDifferences extends RtcmMessageGlonassCorrectionDifferences {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassGeometricCorrectionDifferencesSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGlonassGeometricCorrectionDifferences.prototype, "satellites", void 0);
RtcmMessageGlonassGeometricCorrectionDifferences = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_GEOMETRIC_CORRECTION_DIFFERENCES)
], RtcmMessageGlonassGeometricCorrectionDifferences);
exports.RtcmMessageGlonassGeometricCorrectionDifferences = RtcmMessageGlonassGeometricCorrectionDifferences;
let RtcmMessageGlonassCombinedCorrectionDifferences = class RtcmMessageGlonassCombinedCorrectionDifferences extends RtcmMessageGlonassCorrectionDifferences {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassCombinedCorrectionDifferencesSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGlonassCombinedCorrectionDifferences.prototype, "satellites", void 0);
RtcmMessageGlonassCombinedCorrectionDifferences = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_COMBINED_CORRECTION_DIFFERENCES)
], RtcmMessageGlonassCombinedCorrectionDifferences);
exports.RtcmMessageGlonassCombinedCorrectionDifferences = RtcmMessageGlonassCombinedCorrectionDifferences;
class GlonassNetworkRtkResidualSatelliteData {
}
__decorate([
    decode_encode_1.DF(38),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GlonassNetworkRtkResidualSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(218),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GlonassNetworkRtkResidualSatelliteData.prototype, "s_oc", void 0);
__decorate([
    decode_encode_1.DF(219),
    decode_encode_1.UInt(9),
    __metadata("design:type", Number)
], GlonassNetworkRtkResidualSatelliteData.prototype, "s_od", void 0);
__decorate([
    decode_encode_1.DF(220),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GlonassNetworkRtkResidualSatelliteData.prototype, "s_oh", void 0);
__decorate([
    decode_encode_1.DF(221),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], GlonassNetworkRtkResidualSatelliteData.prototype, "s_ic", void 0);
__decorate([
    decode_encode_1.DF(222),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], GlonassNetworkRtkResidualSatelliteData.prototype, "s_id", void 0);
exports.GlonassNetworkRtkResidualSatelliteData = GlonassNetworkRtkResidualSatelliteData;
let RtcmMessageGlonassNetworkRtkResidual = class RtcmMessageGlonassNetworkRtkResidual extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(225),
    decode_encode_1.UInt(17),
    __metadata("design:type", Number)
], RtcmMessageGlonassNetworkRtkResidual.prototype, "residualsEpochTime", void 0);
__decorate([
    decode_encode_1.DF(3),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGlonassNetworkRtkResidual.prototype, "referenceStationId", void 0);
__decorate([
    decode_encode_1.DF(223),
    decode_encode_1.UInt(7),
    __metadata("design:type", Number)
], RtcmMessageGlonassNetworkRtkResidual.prototype, "nRefs", void 0);
__decorate([
    decode_encode_1.DF(35),
    decode_encode_1.ArrLength(5, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageGlonassNetworkRtkResidual.prototype, "_satelliteCount", void 0);
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassNetworkRtkResidualSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGlonassNetworkRtkResidual.prototype, "satellites", void 0);
RtcmMessageGlonassNetworkRtkResidual = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_NETWORK_RTK_RESIDUAL),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageGlonassNetworkRtkResidual);
exports.RtcmMessageGlonassNetworkRtkResidual = RtcmMessageGlonassNetworkRtkResidual;
class GlonassNetworkFkpGradientSatelliteData {
}
__decorate([
    decode_encode_1.DF(38),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GlonassNetworkFkpGradientSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(392),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GlonassNetworkFkpGradientSatelliteData.prototype, "iode", void 0);
__decorate([
    decode_encode_1.DF(242),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], GlonassNetworkFkpGradientSatelliteData.prototype, "geometricGradientNorth", void 0);
__decorate([
    decode_encode_1.DF(243),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], GlonassNetworkFkpGradientSatelliteData.prototype, "geometricGradientEast", void 0);
__decorate([
    decode_encode_1.DF(244),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], GlonassNetworkFkpGradientSatelliteData.prototype, "ionosphericGradientNorth", void 0);
__decorate([
    decode_encode_1.DF(245),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], GlonassNetworkFkpGradientSatelliteData.prototype, "ionosphericGradientEast", void 0);
exports.GlonassNetworkFkpGradientSatelliteData = GlonassNetworkFkpGradientSatelliteData;
let RtcmMessageGlonassNetworkFkpGradient = class RtcmMessageGlonassNetworkFkpGradient extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(3),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGlonassNetworkFkpGradient.prototype, "referenceStationId", void 0);
__decorate([
    decode_encode_1.DF(241),
    decode_encode_1.UInt(17),
    __metadata("design:type", Number)
], RtcmMessageGlonassNetworkFkpGradient.prototype, "fkpEpochTime", void 0);
__decorate([
    decode_encode_1.DF(35),
    decode_encode_1.ArrLength(5, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageGlonassNetworkFkpGradient.prototype, "_satelliteCount", void 0);
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassNetworkFkpGradientSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGlonassNetworkFkpGradient.prototype, "satellites", void 0);
RtcmMessageGlonassNetworkFkpGradient = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_NETWORK_FKP_GRADIENT),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageGlonassNetworkFkpGradient);
exports.RtcmMessageGlonassNetworkFkpGradient = RtcmMessageGlonassNetworkFkpGradient;
