// @gnss/rtcm/build/messages/network/gps.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmMessageGpsNetworkFkpGradient = exports.GpsNetworkFkpGradientSatelliteData = exports.RtcmMessageGpsNetworkRtkResidual = exports.GpsNetworkRtkResidualSatelliteData = exports.RtcmMessageGpsCombinedCorrectionDifferences = exports.RtcmMessageGpsGeometricCorrectionDifferences = exports.RtcmMessageGpsIonosphericCorrectionDifferences = exports.GpsCombinedCorrectionDifferencesSatelliteData = exports.GpsGeometricCorrectionDifferencesSatelliteData = exports.GpsIonosphericCorrectionDifferencesSatelliteData = exports.GpsCorrectionDifferencesBaseSatelliteData = exports.RtcmMessageGpsCorrectionDifferences = void 0;
const decode_encode_1 = require("../../decode-encode");
const rtcm_1 = require("../../rtcm");
let RtcmMessageGpsCorrectionDifferences = class RtcmMessageGpsCorrectionDifferences extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(59),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], RtcmMessageGpsCorrectionDifferences.prototype, "networkId", void 0);
__decorate([
    decode_encode_1.DF(72),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageGpsCorrectionDifferences.prototype, "subnetworkId", void 0);
__decorate([
    decode_encode_1.DF(65),
    decode_encode_1.UInt(23),
    __metadata("design:type", Number)
], RtcmMessageGpsCorrectionDifferences.prototype, "gpsEpochTime", void 0);
__decorate([
    decode_encode_1.DF(66),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageGpsCorrectionDifferences.prototype, "multipleMessageIndicator", void 0);
__decorate([
    decode_encode_1.DF(60),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGpsCorrectionDifferences.prototype, "masterReferenceStationId", void 0);
__decorate([
    decode_encode_1.DF(61),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGpsCorrectionDifferences.prototype, "auxiliaryReferenceStationId", void 0);
__decorate([
    decode_encode_1.DF(67),
    decode_encode_1.ArrLength(4, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageGpsCorrectionDifferences.prototype, "_satelliteCount", void 0);
RtcmMessageGpsCorrectionDifferences = __decorate([
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageGpsCorrectionDifferences);
exports.RtcmMessageGpsCorrectionDifferences = RtcmMessageGpsCorrectionDifferences;
class GpsCorrectionDifferencesBaseSatelliteData {
}
__decorate([
    decode_encode_1.DF(68),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsCorrectionDifferencesBaseSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(74),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], GpsCorrectionDifferencesBaseSatelliteData.prototype, "ambiguityStatusFlag", void 0);
__decorate([
    decode_encode_1.DF(75),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], GpsCorrectionDifferencesBaseSatelliteData.prototype, "nonSyncCount", void 0);
exports.GpsCorrectionDifferencesBaseSatelliteData = GpsCorrectionDifferencesBaseSatelliteData;
class GpsIonosphericCorrectionDifferencesSatelliteData extends GpsCorrectionDifferencesBaseSatelliteData {
}
__decorate([
    decode_encode_1.DF(69),
    decode_encode_1.Int(17),
    __metadata("design:type", Number)
], GpsIonosphericCorrectionDifferencesSatelliteData.prototype, "ionosphericCarrierPhaseCorrectionDifference", void 0);
exports.GpsIonosphericCorrectionDifferencesSatelliteData = GpsIonosphericCorrectionDifferencesSatelliteData;
class GpsGeometricCorrectionDifferencesSatelliteData extends GpsCorrectionDifferencesBaseSatelliteData {
}
__decorate([
    decode_encode_1.DF(70),
    decode_encode_1.Int(17),
    __metadata("design:type", Number)
], GpsGeometricCorrectionDifferencesSatelliteData.prototype, "geometricCarrierPhaseCorrectionDifference", void 0);
__decorate([
    decode_encode_1.DF(71),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GpsGeometricCorrectionDifferencesSatelliteData.prototype, "iode", void 0);
exports.GpsGeometricCorrectionDifferencesSatelliteData = GpsGeometricCorrectionDifferencesSatelliteData;
class GpsCombinedCorrectionDifferencesSatelliteData extends GpsGeometricCorrectionDifferencesSatelliteData {
}
__decorate([
    decode_encode_1.DF(69),
    decode_encode_1.Int(17),
    __metadata("design:type", Number)
], GpsCombinedCorrectionDifferencesSatelliteData.prototype, "ionosphericCarrierPhaseCorrectionDifference", void 0);
exports.GpsCombinedCorrectionDifferencesSatelliteData = GpsCombinedCorrectionDifferencesSatelliteData;
let RtcmMessageGpsIonosphericCorrectionDifferences = class RtcmMessageGpsIonosphericCorrectionDifferences extends RtcmMessageGpsCorrectionDifferences {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsIonosphericCorrectionDifferencesSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGpsIonosphericCorrectionDifferences.prototype, "satellites", void 0);
RtcmMessageGpsIonosphericCorrectionDifferences = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_IONOSPHERIC_CORRECTION_DIFFERENCES)
], RtcmMessageGpsIonosphericCorrectionDifferences);
exports.RtcmMessageGpsIonosphericCorrectionDifferences = RtcmMessageGpsIonosphericCorrectionDifferences;
let RtcmMessageGpsGeometricCorrectionDifferences = class RtcmMessageGpsGeometricCorrectionDifferences extends RtcmMessageGpsCorrectionDifferences {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsGeometricCorrectionDifferencesSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGpsGeometricCorrectionDifferences.prototype, "satellites", void 0);
RtcmMessageGpsGeometricCorrectionDifferences = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_GEOMETRIC_CORRECTION_DIFFERENCES)
], RtcmMessageGpsGeometricCorrectionDifferences);
exports.RtcmMessageGpsGeometricCorrectionDifferences = RtcmMessageGpsGeometricCorrectionDifferences;
let RtcmMessageGpsCombinedCorrectionDifferences = class RtcmMessageGpsCombinedCorrectionDifferences extends RtcmMessageGpsCorrectionDifferences {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsCombinedCorrectionDifferencesSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGpsCombinedCorrectionDifferences.prototype, "satellites", void 0);
RtcmMessageGpsCombinedCorrectionDifferences = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_COMBINED_CORRECTION_DIFFERENCES)
], RtcmMessageGpsCombinedCorrectionDifferences);
exports.RtcmMessageGpsCombinedCorrectionDifferences = RtcmMessageGpsCombinedCorrectionDifferences;
class GpsNetworkRtkResidualSatelliteData {
}
__decorate([
    decode_encode_1.DF(9),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsNetworkRtkResidualSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(218),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GpsNetworkRtkResidualSatelliteData.prototype, "s_oc", void 0);
__decorate([
    decode_encode_1.DF(219),
    decode_encode_1.UInt(9),
    __metadata("design:type", Number)
], GpsNetworkRtkResidualSatelliteData.prototype, "s_od", void 0);
__decorate([
    decode_encode_1.DF(220),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsNetworkRtkResidualSatelliteData.prototype, "s_oh", void 0);
__decorate([
    decode_encode_1.DF(221),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], GpsNetworkRtkResidualSatelliteData.prototype, "s_ic", void 0);
__decorate([
    decode_encode_1.DF(222),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], GpsNetworkRtkResidualSatelliteData.prototype, "s_id", void 0);
exports.GpsNetworkRtkResidualSatelliteData = GpsNetworkRtkResidualSatelliteData;
let RtcmMessageGpsNetworkRtkResidual = class RtcmMessageGpsNetworkRtkResidual extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(224),
    decode_encode_1.UInt(20),
    __metadata("design:type", Number)
], RtcmMessageGpsNetworkRtkResidual.prototype, "residualsEpochTime", void 0);
__decorate([
    decode_encode_1.DF(3),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGpsNetworkRtkResidual.prototype, "referenceStationId", void 0);
__decorate([
    decode_encode_1.DF(223),
    decode_encode_1.UInt(7),
    __metadata("design:type", Number)
], RtcmMessageGpsNetworkRtkResidual.prototype, "nRefs", void 0);
__decorate([
    decode_encode_1.DF(6),
    decode_encode_1.ArrLength(5, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageGpsNetworkRtkResidual.prototype, "_satelliteCount", void 0);
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsNetworkRtkResidualSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGpsNetworkRtkResidual.prototype, "satellites", void 0);
RtcmMessageGpsNetworkRtkResidual = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_NETWORK_RTK_RESIDUAL),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageGpsNetworkRtkResidual);
exports.RtcmMessageGpsNetworkRtkResidual = RtcmMessageGpsNetworkRtkResidual;
class GpsNetworkFkpGradientSatelliteData {
}
__decorate([
    decode_encode_1.DF(9),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsNetworkFkpGradientSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(71),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GpsNetworkFkpGradientSatelliteData.prototype, "iode", void 0);
__decorate([
    decode_encode_1.DF(242),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], GpsNetworkFkpGradientSatelliteData.prototype, "geometricGradientNorth", void 0);
__decorate([
    decode_encode_1.DF(243),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], GpsNetworkFkpGradientSatelliteData.prototype, "geometricGradientEast", void 0);
__decorate([
    decode_encode_1.DF(244),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], GpsNetworkFkpGradientSatelliteData.prototype, "ionosphericGradientNorth", void 0);
__decorate([
    decode_encode_1.DF(245),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], GpsNetworkFkpGradientSatelliteData.prototype, "ionosphericGradientEast", void 0);
exports.GpsNetworkFkpGradientSatelliteData = GpsNetworkFkpGradientSatelliteData;
let RtcmMessageGpsNetworkFkpGradient = class RtcmMessageGpsNetworkFkpGradient extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(3),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGpsNetworkFkpGradient.prototype, "referenceStationId", void 0);
__decorate([
    decode_encode_1.DF(240),
    decode_encode_1.UInt(20),
    __metadata("design:type", Number)
], RtcmMessageGpsNetworkFkpGradient.prototype, "fkpEpochTime", void 0);
__decorate([
    decode_encode_1.DF(6),
    decode_encode_1.ArrLength(5, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageGpsNetworkFkpGradient.prototype, "_satelliteCount", void 0);
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsNetworkFkpGradientSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageGpsNetworkFkpGradient.prototype, "satellites", void 0);
RtcmMessageGpsNetworkFkpGradient = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_NETWORK_FKP_GRADIENT),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageGpsNetworkFkpGradient);
exports.RtcmMessageGpsNetworkFkpGradient = RtcmMessageGpsNetworkFkpGradient;
