// @gnss/rtcm/build/rtcm.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var RtcmMessage_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmTransport = exports.RtcmMessageUnknown = exports.RtcmMessage = exports.NavSystem = exports.Since = exports.MessageType = exports.RtcmNavSystem = exports.RtcmVersion = exports.RtcmMessageType = void 0;
const bit_buffer_ts_1 = require("bit-buffer-ts");
const utils_1 = require("./utils");
const decode_encode_1 = require("./decode-encode");
const crc_1 = require("./crc");
var RtcmMessageType;
(function (RtcmMessageType) {
    RtcmMessageType[RtcmMessageType["UNKNOWN"] = -1] = "UNKNOWN";
    RtcmMessageType[RtcmMessageType["GPS_L1_OBSERVATIONS"] = 1001] = "GPS_L1_OBSERVATIONS";
    RtcmMessageType[RtcmMessageType["GPS_L1_OBSERVATIONS_EXTENDED"] = 1002] = "GPS_L1_OBSERVATIONS_EXTENDED";
    RtcmMessageType[RtcmMessageType["GPS_L1_L2_OBSERVATIONS"] = 1003] = "GPS_L1_L2_OBSERVATIONS";
    RtcmMessageType[RtcmMessageType["GPS_L1_L2_OBSERVATIONS_EXTENDED"] = 1004] = "GPS_L1_L2_OBSERVATIONS_EXTENDED";
    RtcmMessageType[RtcmMessageType["STATION_ARP"] = 1005] = "STATION_ARP";
    RtcmMessageType[RtcmMessageType["STATION_ARP_HEIGHT"] = 1006] = "STATION_ARP_HEIGHT";
    RtcmMessageType[RtcmMessageType["ANTENNA_DESCRIPTOR"] = 1007] = "ANTENNA_DESCRIPTOR";
    RtcmMessageType[RtcmMessageType["ANTENNA_DESCRIPTOR_SERIAL"] = 1008] = "ANTENNA_DESCRIPTOR_SERIAL";
    RtcmMessageType[RtcmMessageType["GLONASS_L1_OBSERVATIONS"] = 1009] = "GLONASS_L1_OBSERVATIONS";
    RtcmMessageType[RtcmMessageType["GLONASS_L1_OBSERVATIONS_EXTENDED"] = 1010] = "GLONASS_L1_OBSERVATIONS_EXTENDED";
    RtcmMessageType[RtcmMessageType["GLONASS_L1_L2_OBSERVATIONS"] = 1011] = "GLONASS_L1_L2_OBSERVATIONS";
    RtcmMessageType[RtcmMessageType["GLONASS_L1_L2_OBSERVATIONS_EXTENDED"] = 1012] = "GLONASS_L1_L2_OBSERVATIONS_EXTENDED";
    RtcmMessageType[RtcmMessageType["AUXILIARY_OPERATION_INFORMATION"] = 1013] = "AUXILIARY_OPERATION_INFORMATION";
    RtcmMessageType[RtcmMessageType["NETWORK_AUXILIARY_STATION_DATA"] = 1014] = "NETWORK_AUXILIARY_STATION_DATA";
    RtcmMessageType[RtcmMessageType["GPS_IONOSPHERIC_CORRECTION_DIFFERENCES"] = 1015] = "GPS_IONOSPHERIC_CORRECTION_DIFFERENCES";
    RtcmMessageType[RtcmMessageType["GPS_GEOMETRIC_CORRECTION_DIFFERENCES"] = 1016] = "GPS_GEOMETRIC_CORRECTION_DIFFERENCES";
    RtcmMessageType[RtcmMessageType["GPS_COMBINED_CORRECTION_DIFFERENCES"] = 1017] = "GPS_COMBINED_CORRECTION_DIFFERENCES";
    RtcmMessageType[RtcmMessageType["GPS_SATELLITE_EPHEMERIS_DATA"] = 1019] = "GPS_SATELLITE_EPHEMERIS_DATA";
    RtcmMessageType[RtcmMessageType["GLONASS_SATELLITE_EPHEMERIS_DATA"] = 1020] = "GLONASS_SATELLITE_EPHEMERIS_DATA";
    RtcmMessageType[RtcmMessageType["HELMERT_ABRIDGED_MOLODENSKI_TRANSFORMATION_PARAMETERS"] = 1021] = "HELMERT_ABRIDGED_MOLODENSKI_TRANSFORMATION_PARAMETERS";
    RtcmMessageType[RtcmMessageType["MOLODENSKI_BADEKAS_TRANSFORMATION_PARAMETERS"] = 1022] = "MOLODENSKI_BADEKAS_TRANSFORMATION_PARAMETERS";
    RtcmMessageType[RtcmMessageType["RESIDUALS_ELLIPSOIDAL_GRID_REPRESENTATION"] = 1023] = "RESIDUALS_ELLIPSOIDAL_GRID_REPRESENTATION";
    RtcmMessageType[RtcmMessageType["RESIDUALS_PLANE_GRID_REPRESENTATION"] = 1024] = "RESIDUALS_PLANE_GRID_REPRESENTATION";
    RtcmMessageType[RtcmMessageType["PROJECTION_PARAMETERS_EXCEPT_LCC2SP_OM"] = 1025] = "PROJECTION_PARAMETERS_EXCEPT_LCC2SP_OM";
    RtcmMessageType[RtcmMessageType["PROJECTION_PARAMETERS_LCC2SP"] = 1026] = "PROJECTION_PARAMETERS_LCC2SP";
    RtcmMessageType[RtcmMessageType["PROJECTION_PARAMETERS_OM"] = 1027] = "PROJECTION_PARAMETERS_OM";
    RtcmMessageType[RtcmMessageType["UNICODE_TEXT_STRING"] = 1029] = "UNICODE_TEXT_STRING";
    RtcmMessageType[RtcmMessageType["GPS_NETWORK_RTK_RESIDUAL"] = 1030] = "GPS_NETWORK_RTK_RESIDUAL";
    RtcmMessageType[RtcmMessageType["GLONASS_NETWORK_RTK_RESIDUAL"] = 1031] = "GLONASS_NETWORK_RTK_RESIDUAL";
    RtcmMessageType[RtcmMessageType["PHYSICAL_REFERENCE_STATION_POSITION"] = 1032] = "PHYSICAL_REFERENCE_STATION_POSITION";
    RtcmMessageType[RtcmMessageType["RECEIVER_ANTENNA_DESCRIPTOR"] = 1033] = "RECEIVER_ANTENNA_DESCRIPTOR";
    RtcmMessageType[RtcmMessageType["GPS_NETWORK_FKP_GRADIENT"] = 1034] = "GPS_NETWORK_FKP_GRADIENT";
    RtcmMessageType[RtcmMessageType["GLONASS_NETWORK_FKP_GRADIENT"] = 1035] = "GLONASS_NETWORK_FKP_GRADIENT";
    RtcmMessageType[RtcmMessageType["GLONASS_IONOSPHERIC_CORRECTION_DIFFERENCES"] = 1037] = "GLONASS_IONOSPHERIC_CORRECTION_DIFFERENCES";
    RtcmMessageType[RtcmMessageType["GLONASS_GEOMETRIC_CORRECTION_DIFFERENCES"] = 1038] = "GLONASS_GEOMETRIC_CORRECTION_DIFFERENCES";
    RtcmMessageType[RtcmMessageType["GLONASS_COMBINED_CORRECTION_DIFFERENCES"] = 1039] = "GLONASS_COMBINED_CORRECTION_DIFFERENCES";
    RtcmMessageType[RtcmMessageType["IRNSS_SATELLITE_EPHEMERIS_DATA"] = 1041] = "IRNSS_SATELLITE_EPHEMERIS_DATA";
    RtcmMessageType[RtcmMessageType["BEIDOU_SATELLITE_EPHEMERIS_DATA"] = 1042] = "BEIDOU_SATELLITE_EPHEMERIS_DATA";
    RtcmMessageType[RtcmMessageType["QZSS_SATELLITE_EPHEMERIS_DATA"] = 1044] = "QZSS_SATELLITE_EPHEMERIS_DATA";
    RtcmMessageType[RtcmMessageType["GALILEO_F_NAV_SATELLITE_EPHEMERIS_DATA"] = 1045] = "GALILEO_F_NAV_SATELLITE_EPHEMERIS_DATA";
    RtcmMessageType[RtcmMessageType["GALILEO_I_NAV_SATELLITE_EPHEMERIS_DATA"] = 1046] = "GALILEO_I_NAV_SATELLITE_EPHEMERIS_DATA";
    RtcmMessageType[RtcmMessageType["SSR_GPS_ORBIT_CORRECTION"] = 1057] = "SSR_GPS_ORBIT_CORRECTION";
    RtcmMessageType[RtcmMessageType["SSR_GPS_CLOCK_CORRECTION"] = 1058] = "SSR_GPS_CLOCK_CORRECTION";
    RtcmMessageType[RtcmMessageType["SSR_GPS_CODE_BIAS"] = 1059] = "SSR_GPS_CODE_BIAS";
    RtcmMessageType[RtcmMessageType["SSR_GPS_COMBINED_CORRECTION"] = 1060] = "SSR_GPS_COMBINED_CORRECTION";
    RtcmMessageType[RtcmMessageType["SSR_GPS_URA"] = 1061] = "SSR_GPS_URA";
    RtcmMessageType[RtcmMessageType["SSR_GPS_HIGH_RATE_CLOCK_CORRECTION"] = 1062] = "SSR_GPS_HIGH_RATE_CLOCK_CORRECTION";
    RtcmMessageType[RtcmMessageType["SSR_GLONASS_ORBIT_CORRECTION"] = 1063] = "SSR_GLONASS_ORBIT_CORRECTION";
    RtcmMessageType[RtcmMessageType["SSR_GLONASS_CLOCK_CORRECTION"] = 1064] = "SSR_GLONASS_CLOCK_CORRECTION";
    RtcmMessageType[RtcmMessageType["SSR_GLONASS_CODE_BIAS"] = 1065] = "SSR_GLONASS_CODE_BIAS";
    RtcmMessageType[RtcmMessageType["SSR_GLONASS_COMBINED_CORRECTION"] = 1066] = "SSR_GLONASS_COMBINED_CORRECTION";
    RtcmMessageType[RtcmMessageType["SSR_GLONASS_URA"] = 1067] = "SSR_GLONASS_URA";
    RtcmMessageType[RtcmMessageType["SSR_GLONASS_HIGH_RATE_CLOCK_CORRECTION"] = 1068] = "SSR_GLONASS_HIGH_RATE_CLOCK_CORRECTION";
    RtcmMessageType[RtcmMessageType["GPS_MSM1"] = 1071] = "GPS_MSM1";
    RtcmMessageType[RtcmMessageType["GPS_MSM2"] = 1072] = "GPS_MSM2";
    RtcmMessageType[RtcmMessageType["GPS_MSM3"] = 1073] = "GPS_MSM3";
    RtcmMessageType[RtcmMessageType["GPS_MSM4"] = 1074] = "GPS_MSM4";
    RtcmMessageType[RtcmMessageType["GPS_MSM5"] = 1075] = "GPS_MSM5";
    RtcmMessageType[RtcmMessageType["GPS_MSM6"] = 1076] = "GPS_MSM6";
    RtcmMessageType[RtcmMessageType["GPS_MSM7"] = 1077] = "GPS_MSM7";
    RtcmMessageType[RtcmMessageType["GLONASS_MSM1"] = 1081] = "GLONASS_MSM1";
    RtcmMessageType[RtcmMessageType["GLONASS_MSM2"] = 1082] = "GLONASS_MSM2";
    RtcmMessageType[RtcmMessageType["GLONASS_MSM3"] = 1083] = "GLONASS_MSM3";
    RtcmMessageType[RtcmMessageType["GLONASS_MSM4"] = 1084] = "GLONASS_MSM4";
    RtcmMessageType[RtcmMessageType["GLONASS_MSM5"] = 1085] = "GLONASS_MSM5";
    RtcmMessageType[RtcmMessageType["GLONASS_MSM6"] = 1086] = "GLONASS_MSM6";
    RtcmMessageType[RtcmMessageType["GLONASS_MSM7"] = 1087] = "GLONASS_MSM7";
    RtcmMessageType[RtcmMessageType["GALILEO_MSM1"] = 1091] = "GALILEO_MSM1";
    RtcmMessageType[RtcmMessageType["GALILEO_MSM2"] = 1092] = "GALILEO_MSM2";
    RtcmMessageType[RtcmMessageType["GALILEO_MSM3"] = 1093] = "GALILEO_MSM3";
    RtcmMessageType[RtcmMessageType["GALILEO_MSM4"] = 1094] = "GALILEO_MSM4";
    RtcmMessageType[RtcmMessageType["GALILEO_MSM5"] = 1095] = "GALILEO_MSM5";
    RtcmMessageType[RtcmMessageType["GALILEO_MSM6"] = 1096] = "GALILEO_MSM6";
    RtcmMessageType[RtcmMessageType["GALILEO_MSM7"] = 1097] = "GALILEO_MSM7";
    RtcmMessageType[RtcmMessageType["SBAS_MSM1"] = 1101] = "SBAS_MSM1";
    RtcmMessageType[RtcmMessageType["SBAS_MSM2"] = 1102] = "SBAS_MSM2";
    RtcmMessageType[RtcmMessageType["SBAS_MSM3"] = 1103] = "SBAS_MSM3";
    RtcmMessageType[RtcmMessageType["SBAS_MSM4"] = 1104] = "SBAS_MSM4";
    RtcmMessageType[RtcmMessageType["SBAS_MSM5"] = 1105] = "SBAS_MSM5";
    RtcmMessageType[RtcmMessageType["SBAS_MSM6"] = 1106] = "SBAS_MSM6";
    RtcmMessageType[RtcmMessageType["SBAS_MSM7"] = 1107] = "SBAS_MSM7";
    RtcmMessageType[RtcmMessageType["QZSS_MSM1"] = 1111] = "QZSS_MSM1";
    RtcmMessageType[RtcmMessageType["QZSS_MSM2"] = 1112] = "QZSS_MSM2";
    RtcmMessageType[RtcmMessageType["QZSS_MSM3"] = 1113] = "QZSS_MSM3";
    RtcmMessageType[RtcmMessageType["QZSS_MSM4"] = 1114] = "QZSS_MSM4";
    RtcmMessageType[RtcmMessageType["QZSS_MSM5"] = 1115] = "QZSS_MSM5";
    RtcmMessageType[RtcmMessageType["QZSS_MSM6"] = 1116] = "QZSS_MSM6";
    RtcmMessageType[RtcmMessageType["QZSS_MSM7"] = 1117] = "QZSS_MSM7";
    RtcmMessageType[RtcmMessageType["BEIDOU_MSM1"] = 1121] = "BEIDOU_MSM1";
    RtcmMessageType[RtcmMessageType["BEIDOU_MSM2"] = 1122] = "BEIDOU_MSM2";
    RtcmMessageType[RtcmMessageType["BEIDOU_MSM3"] = 1123] = "BEIDOU_MSM3";
    RtcmMessageType[RtcmMessageType["BEIDOU_MSM4"] = 1124] = "BEIDOU_MSM4";
    RtcmMessageType[RtcmMessageType["BEIDOU_MSM5"] = 1125] = "BEIDOU_MSM5";
    RtcmMessageType[RtcmMessageType["BEIDOU_MSM6"] = 1126] = "BEIDOU_MSM6";
    RtcmMessageType[RtcmMessageType["BEIDOU_MSM7"] = 1127] = "BEIDOU_MSM7";
    RtcmMessageType[RtcmMessageType["IRNSS_MSM1"] = 1131] = "IRNSS_MSM1";
    RtcmMessageType[RtcmMessageType["IRNSS_MSM2"] = 1132] = "IRNSS_MSM2";
    RtcmMessageType[RtcmMessageType["IRNSS_MSM3"] = 1133] = "IRNSS_MSM3";
    RtcmMessageType[RtcmMessageType["IRNSS_MSM4"] = 1134] = "IRNSS_MSM4";
    RtcmMessageType[RtcmMessageType["IRNSS_MSM5"] = 1135] = "IRNSS_MSM5";
    RtcmMessageType[RtcmMessageType["IRNSS_MSM6"] = 1136] = "IRNSS_MSM6";
    RtcmMessageType[RtcmMessageType["IRNSS_MSM7"] = 1137] = "IRNSS_MSM7";
    RtcmMessageType[RtcmMessageType["FUTURE_1_MSM1"] = 1141] = "FUTURE_1_MSM1";
    RtcmMessageType[RtcmMessageType["FUTURE_1_MSM2"] = 1142] = "FUTURE_1_MSM2";
    RtcmMessageType[RtcmMessageType["FUTURE_1_MSM3"] = 1143] = "FUTURE_1_MSM3";
    RtcmMessageType[RtcmMessageType["FUTURE_1_MSM4"] = 1144] = "FUTURE_1_MSM4";
    RtcmMessageType[RtcmMessageType["FUTURE_1_MSM5"] = 1145] = "FUTURE_1_MSM5";
    RtcmMessageType[RtcmMessageType["FUTURE_1_MSM6"] = 1146] = "FUTURE_1_MSM6";
    RtcmMessageType[RtcmMessageType["FUTURE_1_MSM7"] = 1147] = "FUTURE_1_MSM7";
    RtcmMessageType[RtcmMessageType["FUTURE_2_MSM1"] = 1151] = "FUTURE_2_MSM1";
    RtcmMessageType[RtcmMessageType["FUTURE_2_MSM2"] = 1152] = "FUTURE_2_MSM2";
    RtcmMessageType[RtcmMessageType["FUTURE_2_MSM3"] = 1153] = "FUTURE_2_MSM3";
    RtcmMessageType[RtcmMessageType["FUTURE_2_MSM4"] = 1154] = "FUTURE_2_MSM4";
    RtcmMessageType[RtcmMessageType["FUTURE_2_MSM5"] = 1155] = "FUTURE_2_MSM5";
    RtcmMessageType[RtcmMessageType["FUTURE_2_MSM6"] = 1156] = "FUTURE_2_MSM6";
    RtcmMessageType[RtcmMessageType["FUTURE_2_MSM7"] = 1157] = "FUTURE_2_MSM7";
    RtcmMessageType[RtcmMessageType["FUTURE_3_MSM1"] = 1161] = "FUTURE_3_MSM1";
    RtcmMessageType[RtcmMessageType["FUTURE_3_MSM2"] = 1162] = "FUTURE_3_MSM2";
    RtcmMessageType[RtcmMessageType["FUTURE_3_MSM3"] = 1163] = "FUTURE_3_MSM3";
    RtcmMessageType[RtcmMessageType["FUTURE_3_MSM4"] = 1164] = "FUTURE_3_MSM4";
    RtcmMessageType[RtcmMessageType["FUTURE_3_MSM5"] = 1165] = "FUTURE_3_MSM5";
    RtcmMessageType[RtcmMessageType["FUTURE_3_MSM6"] = 1166] = "FUTURE_3_MSM6";
    RtcmMessageType[RtcmMessageType["FUTURE_3_MSM7"] = 1167] = "FUTURE_3_MSM7";
    RtcmMessageType[RtcmMessageType["FUTURE_4_MSM1"] = 1171] = "FUTURE_4_MSM1";
    RtcmMessageType[RtcmMessageType["FUTURE_4_MSM2"] = 1172] = "FUTURE_4_MSM2";
    RtcmMessageType[RtcmMessageType["FUTURE_4_MSM3"] = 1173] = "FUTURE_4_MSM3";
    RtcmMessageType[RtcmMessageType["FUTURE_4_MSM4"] = 1174] = "FUTURE_4_MSM4";
    RtcmMessageType[RtcmMessageType["FUTURE_4_MSM5"] = 1175] = "FUTURE_4_MSM5";
    RtcmMessageType[RtcmMessageType["FUTURE_4_MSM6"] = 1176] = "FUTURE_4_MSM6";
    RtcmMessageType[RtcmMessageType["FUTURE_4_MSM7"] = 1177] = "FUTURE_4_MSM7";
    RtcmMessageType[RtcmMessageType["FUTURE_5_MSM1"] = 1181] = "FUTURE_5_MSM1";
    RtcmMessageType[RtcmMessageType["FUTURE_5_MSM2"] = 1182] = "FUTURE_5_MSM2";
    RtcmMessageType[RtcmMessageType["FUTURE_5_MSM3"] = 1183] = "FUTURE_5_MSM3";
    RtcmMessageType[RtcmMessageType["FUTURE_5_MSM4"] = 1184] = "FUTURE_5_MSM4";
    RtcmMessageType[RtcmMessageType["FUTURE_5_MSM5"] = 1185] = "FUTURE_5_MSM5";
    RtcmMessageType[RtcmMessageType["FUTURE_5_MSM6"] = 1186] = "FUTURE_5_MSM6";
    RtcmMessageType[RtcmMessageType["FUTURE_5_MSM7"] = 1187] = "FUTURE_5_MSM7";
    RtcmMessageType[RtcmMessageType["FUTURE_6_MSM1"] = 1191] = "FUTURE_6_MSM1";
    RtcmMessageType[RtcmMessageType["FUTURE_6_MSM2"] = 1192] = "FUTURE_6_MSM2";
    RtcmMessageType[RtcmMessageType["FUTURE_6_MSM3"] = 1193] = "FUTURE_6_MSM3";
    RtcmMessageType[RtcmMessageType["FUTURE_6_MSM4"] = 1194] = "FUTURE_6_MSM4";
    RtcmMessageType[RtcmMessageType["FUTURE_6_MSM5"] = 1195] = "FUTURE_6_MSM5";
    RtcmMessageType[RtcmMessageType["FUTURE_6_MSM6"] = 1196] = "FUTURE_6_MSM6";
    RtcmMessageType[RtcmMessageType["FUTURE_6_MSM7"] = 1197] = "FUTURE_6_MSM7";
    RtcmMessageType[RtcmMessageType["FUTURE_7_MSM1"] = 1201] = "FUTURE_7_MSM1";
    RtcmMessageType[RtcmMessageType["FUTURE_7_MSM2"] = 1202] = "FUTURE_7_MSM2";
    RtcmMessageType[RtcmMessageType["FUTURE_7_MSM3"] = 1203] = "FUTURE_7_MSM3";
    RtcmMessageType[RtcmMessageType["FUTURE_7_MSM4"] = 1204] = "FUTURE_7_MSM4";
    RtcmMessageType[RtcmMessageType["FUTURE_7_MSM5"] = 1205] = "FUTURE_7_MSM5";
    RtcmMessageType[RtcmMessageType["FUTURE_7_MSM6"] = 1206] = "FUTURE_7_MSM6";
    RtcmMessageType[RtcmMessageType["FUTURE_7_MSM7"] = 1207] = "FUTURE_7_MSM7";
    RtcmMessageType[RtcmMessageType["FUTURE_8_MSM1"] = 1211] = "FUTURE_8_MSM1";
    RtcmMessageType[RtcmMessageType["FUTURE_8_MSM2"] = 1212] = "FUTURE_8_MSM2";
    RtcmMessageType[RtcmMessageType["FUTURE_8_MSM3"] = 1213] = "FUTURE_8_MSM3";
    RtcmMessageType[RtcmMessageType["FUTURE_8_MSM4"] = 1214] = "FUTURE_8_MSM4";
    RtcmMessageType[RtcmMessageType["FUTURE_8_MSM5"] = 1215] = "FUTURE_8_MSM5";
    RtcmMessageType[RtcmMessageType["FUTURE_8_MSM6"] = 1216] = "FUTURE_8_MSM6";
    RtcmMessageType[RtcmMessageType["FUTURE_8_MSM7"] = 1217] = "FUTURE_8_MSM7";
    RtcmMessageType[RtcmMessageType["FUTURE_9_MSM1"] = 1221] = "FUTURE_9_MSM1";
    RtcmMessageType[RtcmMessageType["FUTURE_9_MSM2"] = 1222] = "FUTURE_9_MSM2";
    RtcmMessageType[RtcmMessageType["FUTURE_9_MSM3"] = 1223] = "FUTURE_9_MSM3";
    RtcmMessageType[RtcmMessageType["FUTURE_9_MSM4"] = 1224] = "FUTURE_9_MSM4";
    RtcmMessageType[RtcmMessageType["FUTURE_9_MSM5"] = 1225] = "FUTURE_9_MSM5";
    RtcmMessageType[RtcmMessageType["FUTURE_9_MSM6"] = 1226] = "FUTURE_9_MSM6";
    RtcmMessageType[RtcmMessageType["FUTURE_9_MSM7"] = 1227] = "FUTURE_9_MSM7";
    RtcmMessageType[RtcmMessageType["GLONASS_L1_L2_CODE_PHASE_BIASES"] = 1230] = "GLONASS_L1_L2_CODE_PHASE_BIASES";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_ASHTECH"] = 4095] = "PROPRIETARY_ASHTECH";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_TRIMBLE"] = 4094] = "PROPRIETARY_TRIMBLE";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_NOVATEL"] = 4093] = "PROPRIETARY_NOVATEL";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_LEICA"] = 4092] = "PROPRIETARY_LEICA";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_TOPCON"] = 4091] = "PROPRIETARY_TOPCON";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_GEO"] = 4090] = "PROPRIETARY_GEO";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_SEPTENTRIO"] = 4089] = "PROPRIETARY_SEPTENTRIO";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_IFEN"] = 4088] = "PROPRIETARY_IFEN";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_FUGRO"] = 4087] = "PROPRIETARY_FUGRO";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_INPOSITION"] = 4086] = "PROPRIETARY_INPOSITION";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_GSA"] = 4085] = "PROPRIETARY_GSA";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_GEODETICS"] = 4084] = "PROPRIETARY_GEODETICS";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_DLR"] = 4083] = "PROPRIETARY_DLR";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_CRCSI"] = 4082] = "PROPRIETARY_CRCSI";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_SNUGL"] = 4081] = "PROPRIETARY_SNUGL";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_NAVCOM"] = 4080] = "PROPRIETARY_NAVCOM";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_SCSC"] = 4079] = "PROPRIETARY_SCSC";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_COMNAV"] = 4078] = "PROPRIETARY_COMNAV";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_HEMISPHERE"] = 4077] = "PROPRIETARY_HEMISPHERE";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_IGS"] = 4076] = "PROPRIETARY_IGS";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_ALBERDING"] = 4075] = "PROPRIETARY_ALBERDING";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_UNICORE"] = 4074] = "PROPRIETARY_UNICORE";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_MITSUBISHI"] = 4073] = "PROPRIETARY_MITSUBISHI";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_UBLOX"] = 4072] = "PROPRIETARY_UBLOX";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_WNLBS"] = 4071] = "PROPRIETARY_WNLBS";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_MENGXIN"] = 4070] = "PROPRIETARY_MENGXIN";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_VERIPOS"] = 4069] = "PROPRIETARY_VERIPOS";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_QIANXUN"] = 4068] = "PROPRIETARY_QIANXUN";
    RtcmMessageType[RtcmMessageType["PROPRIETARY_CTTIC"] = 4067] = "PROPRIETARY_CTTIC";
})(RtcmMessageType = exports.RtcmMessageType || (exports.RtcmMessageType = {}));
var RtcmVersion;
(function (RtcmVersion) {
    RtcmVersion[RtcmVersion["V3_0"] = 3] = "V3_0";
    RtcmVersion[RtcmVersion["V3_1"] = 3.1] = "V3_1";
    RtcmVersion[RtcmVersion["V3_2"] = 3.2] = "V3_2";
    RtcmVersion[RtcmVersion["V3_3"] = 3.3] = "V3_3";
    RtcmVersion[RtcmVersion["FUTURE"] = 4] = "FUTURE";
})(RtcmVersion = exports.RtcmVersion || (exports.RtcmVersion = {}));
var RtcmNavSystem;
(function (RtcmNavSystem) {
    RtcmNavSystem["GPS"] = "GPS";
    RtcmNavSystem["GLONASS"] = "GLONASS";
    RtcmNavSystem["GALILEO"] = "GALILEO";
    RtcmNavSystem["QZSS"] = "QZSS";
    RtcmNavSystem["SBAS"] = "SBAS";
    RtcmNavSystem["BEIDOU"] = "BEIDOU";
    RtcmNavSystem["IRNSS"] = "IRNSS";
    RtcmNavSystem["FUTURE"] = "FUTURE";
})(RtcmNavSystem = exports.RtcmNavSystem || (exports.RtcmNavSystem = {}));
const rtcmMessageClasses = new Map();
const MessageType = (type) => (constructor) => {
    Object.defineProperty(constructor, 'messageType', {
        value: type,
        writable: false
    });
    if (rtcmMessageClasses.has(type))
        throw new Error(`Class for message type ${type} is already registered`);
    rtcmMessageClasses.set(type, constructor);
};
exports.MessageType = MessageType;
const sinceMetadataKey = Symbol('since');
const Since = (version) => Reflect.metadata(sinceMetadataKey, version);
exports.Since = Since;
const navSystemMetadataKey = Symbol('navSystem');
const NavSystem = (version) => Reflect.metadata(navSystemMetadataKey, version);
exports.NavSystem = NavSystem;
let RtcmMessage = RtcmMessage_1 = class RtcmMessage {
    constructor(internalGuard) { }
    static get decoderEncoder() { return decode_encode_1.getDecoderEncoder(this); }
    static get sinceVersion() { return Reflect.getMetadata(sinceMetadataKey, this); }
    static get navSystem() { var _a; return (_a = Reflect.getMetadata(navSystemMetadataKey, this)) !== null && _a !== void 0 ? _a : null; }
    get messageType() { return this.constructor.messageType; }
    ;
    static construct(params) {
        return Object.assign(new this(undefined), params);
    }
};
RtcmMessage.messageType = RtcmMessageType.UNKNOWN;
RtcmMessage = RtcmMessage_1 = __decorate([
    exports.Since(RtcmVersion.V3_0),
    __metadata("design:paramtypes", [void 0])
], RtcmMessage);
exports.RtcmMessage = RtcmMessage;
let RtcmMessageUnknown = class RtcmMessageUnknown extends RtcmMessage {
    get messageType() {
        return this.raw.length >= 2 ? (this.raw[0] << 4 || this.raw[1] >> 4) : RtcmMessageType.UNKNOWN;
    }
};
RtcmMessageUnknown = __decorate([
    exports.Since(RtcmVersion.V3_0)
], RtcmMessageUnknown);
exports.RtcmMessageUnknown = RtcmMessageUnknown;
var RtcmTransport;
(function (RtcmTransport) {
    RtcmTransport.SYNC_CHAR = 0xD3;
    RtcmTransport.HEADER_SIZE = 3;
    RtcmTransport.CRC_SIZE = 3;
    RtcmTransport.MAX_PACKET_SIZE = 1023 + RtcmTransport.HEADER_SIZE + RtcmTransport.CRC_SIZE;
    class Exception extends Error {
        constructor(type, message) {
            super(`RTCM ${type} exception: ${message}`);
        }
    }
    RtcmTransport.Exception = Exception;
    class DecodeException extends Exception {
        constructor(message) {
            super('decode', message);
        }
    }
    RtcmTransport.DecodeException = DecodeException;
    class EncodeException extends Exception {
        constructor(message) {
            super('encode', message);
        }
    }
    RtcmTransport.EncodeException = EncodeException;
    function readUint(buffer, offset, byteLength) {
        let value = 0;
        for (let i = 0; i < byteLength; i++) {
            value = value << 8 | buffer[offset + i];
        }
        return value;
    }
    function writeUint(buffer, value, offset, byteLength) {
        for (let i = byteLength - 1; i >= 0; i--, value >>= 8) {
            buffer[offset + i] = value & 0xff;
        }
    }
    function decode(buffer) {
        const preamble = buffer[0];
        if (preamble != RtcmTransport.SYNC_CHAR)
            throw new DecodeException(`Invalid preamble (expected ${RtcmTransport.SYNC_CHAR.toString(16)}, got ${preamble.toString(16)})`);
        const messageLength = (buffer[1] & 0x03) << 8 | buffer[2];
        const messageEndByteIndex = RtcmTransport.HEADER_SIZE + messageLength;
        const crcContent = buffer.slice(0, messageEndByteIndex);
        const crcCalculated = crc24q(crcContent);
        const crcReceived = readUint(buffer, messageEndByteIndex, RtcmTransport.CRC_SIZE);
        if (crcCalculated != crcReceived)
            throw new DecodeException(`CRC does not match (expected ${crcReceived.toString(16)}, got ${crcCalculated.toString(16)})`);
        const messageType = messageLength >= 2 ? (buffer[3] << 4 | buffer[4] >> 4) : RtcmMessageType.UNKNOWN;
        let message;
        const resultType = rtcmMessageClasses.get(messageType);
        if (resultType === undefined) {
            const messageContent = buffer.slice(RtcmTransport.HEADER_SIZE, messageEndByteIndex);
            message = RtcmMessageUnknown.construct({
                raw: messageContent
            });
        }
        else {
            message = new resultType(undefined);
            const s = new bit_buffer_ts_1.BitStream(buffer, RtcmTransport.HEADER_SIZE);
            s.index += 12;
            const decoder = decode_encode_1.getDecoderEncoder(message.constructor).decoder;
            try {
                decoder.run(message, s);
            }
            catch (err) {
                throw new DecodeException(`Could not run script: ${err.message}\n${decoder.fullScript}\n${JSON.stringify(message)}`);
            }
            const finalByteIndex = s.byteIndex;
            if (finalByteIndex > messageLength)
                throw new DecodeException(`Corrupt ${resultType.name}[${messageType}] detected, not long enough to contain necessary data (${messageLength} sent, ${finalByteIndex} required)`);
        }
        return [message, messageLength + RtcmTransport.HEADER_SIZE + RtcmTransport.CRC_SIZE];
    }
    RtcmTransport.decode = decode;
    function encode(message, buffer) {
        buffer[0] = RtcmTransport.SYNC_CHAR;
        let messageLength;
        if (message instanceof RtcmMessageUnknown) {
            messageLength = message.raw.length;
            message.raw.set(buffer, RtcmTransport.HEADER_SIZE);
        }
        else {
            const s = new bit_buffer_ts_1.BitStream(buffer, RtcmTransport.HEADER_SIZE);
            s.writeBits(message.messageType, 12);
            const encoder = decode_encode_1.getDecoderEncoder(message.constructor).encoder;
            try {
                encoder.run(message, s);
            }
            catch (err) {
                throw new EncodeException(`Could not run script: ${err.message}\n${encoder.fullScript}`);
            }
            messageLength = s.byteIndex;
        }
        writeUint(buffer, messageLength, 1, 2);
        const messageEndByteIndex = RtcmTransport.HEADER_SIZE + messageLength;
        const crcContent = buffer.slice(0, messageEndByteIndex);
        writeUint(buffer, crc24q(crcContent), messageEndByteIndex, RtcmTransport.CRC_SIZE);
        return messageLength + RtcmTransport.HEADER_SIZE + RtcmTransport.CRC_SIZE;
    }
    RtcmTransport.encode = encode;
    function crc24q(buffer) {
        let crc = 0;
        for (let i = 0; i < buffer.length; i++) {
            crc = (crc << 8) ^ crc_1.CRC24Q_TABLE[(buffer[i] ^ (crc >> 16)) & 0xff];
        }
        return crc & 0xffffff;
    }
    RtcmTransport.crc24q = crc24q;
})(RtcmTransport = exports.RtcmTransport || (exports.RtcmTransport = {}));
require("./messages/index");
