// @gnss/rtcm/build/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constructWithPropertiesAndParametersPlus = exports.constructWithPropertiesAndParameters = exports.constructWithPropertiesInternalGuardPlus = exports.constructWithPropertiesInternalGuard = exports.constructWithPropertiesPlus = exports.constructWithProperties = exports.constructPropertiesKey = void 0;
exports.constructPropertiesKey = Symbol('constructProperties');
function constructWithProperties(properties) {
    return Object.assign(new this(), properties);
}
exports.constructWithProperties = constructWithProperties;
function constructWithPropertiesPlus(properties) {
    return Object.assign(new this(), properties);
}
exports.constructWithPropertiesPlus = constructWithPropertiesPlus;
function constructWithPropertiesInternalGuard(properties) {
    return Object.assign(new this(undefined), properties);
}
exports.constructWithPropertiesInternalGuard = constructWithPropertiesInternalGuard;
function constructWithPropertiesInternalGuardPlus(properties) {
    return Object.assign(new this(undefined), properties);
}
exports.constructWithPropertiesInternalGuardPlus = constructWithPropertiesInternalGuardPlus;
function constructWithPropertiesAndParameters(properties, ...parameters) {
    return Object.assign(new this(...parameters), properties);
}
exports.constructWithPropertiesAndParameters = constructWithPropertiesAndParameters;
function constructWithPropertiesAndParametersPlus(properties, ...parameters) {
    return Object.assign(new this(...parameters), properties);
}
exports.constructWithPropertiesAndParametersPlus = constructWithPropertiesAndParametersPlus;
class Test {
    constructor() {
        this.x = 1;
    }
}
let x;
let y;
