// @gnss/rtcm/build/messages/observations/msm/gnss.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmMessageMsm1Future1 = exports.RtcmMessageMsm7Irnss = exports.RtcmMessageMsm6Irnss = exports.RtcmMessageMsm5Irnss = exports.RtcmMessageMsm4Irnss = exports.RtcmMessageMsm3Irnss = exports.RtcmMessageMsm2Irnss = exports.RtcmMessageMsm1Irnss = exports.RtcmMessageMsm7Bds = exports.RtcmMessageMsm6Bds = exports.RtcmMessageMsm5Bds = exports.RtcmMessageMsm4Bds = exports.RtcmMessageMsm3Bds = exports.RtcmMessageMsm2Bds = exports.RtcmMessageMsm1Bds = exports.RtcmMessageMsm7Qzss = exports.RtcmMessageMsm6Qzss = exports.RtcmMessageMsm5Qzss = exports.RtcmMessageMsm4Qzss = exports.RtcmMessageMsm3Qzss = exports.RtcmMessageMsm2Qzss = exports.RtcmMessageMsm1Qzss = exports.RtcmMessageMsm7Sbas = exports.RtcmMessageMsm6Sbas = exports.RtcmMessageMsm5Sbas = exports.RtcmMessageMsm4Sbas = exports.RtcmMessageMsm3Sbas = exports.RtcmMessageMsm2Sbas = exports.RtcmMessageMsm1Sbas = exports.RtcmMessageMsm7Galileo = exports.RtcmMessageMsm6Galileo = exports.RtcmMessageMsm5Galileo = exports.RtcmMessageMsm4Galileo = exports.RtcmMessageMsm3Galileo = exports.RtcmMessageMsm2Galileo = exports.RtcmMessageMsm1Galileo = exports.RtcmMessageMsm7Glonass = exports.RtcmMessageMsm6Glonass = exports.RtcmMessageMsm5Glonass = exports.RtcmMessageMsm4Glonass = exports.RtcmMessageMsm3Glonass = exports.RtcmMessageMsm2Glonass = exports.RtcmMessageMsm1Glonass = exports.RtcmMessageMsm7Gps = exports.RtcmMessageMsm6Gps = exports.RtcmMessageMsm5Gps = exports.RtcmMessageMsm4Gps = exports.RtcmMessageMsm3Gps = exports.RtcmMessageMsm2Gps = exports.RtcmMessageMsm1Gps = void 0;
exports.RtcmMessageMsm2Future8 = exports.RtcmMessageMsm1Future8 = exports.RtcmMessageMsm7Future7 = exports.RtcmMessageMsm6Future7 = exports.RtcmMessageMsm5Future7 = exports.RtcmMessageMsm4Future7 = exports.RtcmMessageMsm3Future7 = exports.RtcmMessageMsm2Future7 = exports.RtcmMessageMsm1Future7 = exports.RtcmMessageMsm7Future6 = exports.RtcmMessageMsm6Future6 = exports.RtcmMessageMsm5Future6 = exports.RtcmMessageMsm4Future6 = exports.RtcmMessageMsm3Future6 = exports.RtcmMessageMsm2Future6 = exports.RtcmMessageMsm1Future6 = exports.RtcmMessageMsm7Future5 = exports.RtcmMessageMsm6Future5 = exports.RtcmMessageMsm5Future5 = exports.RtcmMessageMsm4Future5 = exports.RtcmMessageMsm3Future5 = exports.RtcmMessageMsm2Future5 = exports.RtcmMessageMsm1Future5 = exports.RtcmMessageMsm7Future4 = exports.RtcmMessageMsm6Future4 = exports.RtcmMessageMsm5Future4 = exports.RtcmMessageMsm4Future4 = exports.RtcmMessageMsm3Future4 = exports.RtcmMessageMsm2Future4 = exports.RtcmMessageMsm1Future4 = exports.RtcmMessageMsm7Future3 = exports.RtcmMessageMsm6Future3 = exports.RtcmMessageMsm5Future3 = exports.RtcmMessageMsm4Future3 = exports.RtcmMessageMsm3Future3 = exports.RtcmMessageMsm2Future3 = exports.RtcmMessageMsm1Future3 = exports.RtcmMessageMsm7Future2 = exports.RtcmMessageMsm6Future2 = exports.RtcmMessageMsm5Future2 = exports.RtcmMessageMsm4Future2 = exports.RtcmMessageMsm3Future2 = exports.RtcmMessageMsm2Future2 = exports.RtcmMessageMsm1Future2 = exports.RtcmMessageMsm7Future1 = exports.RtcmMessageMsm6Future1 = exports.RtcmMessageMsm5Future1 = exports.RtcmMessageMsm4Future1 = exports.RtcmMessageMsm3Future1 = exports.RtcmMessageMsm2Future1 = void 0;
exports.signalIdMapping = exports.referenceSignals = exports.RtcmMessageMsm7Future9 = exports.RtcmMessageMsm6Future9 = exports.RtcmMessageMsm5Future9 = exports.RtcmMessageMsm4Future9 = exports.RtcmMessageMsm3Future9 = exports.RtcmMessageMsm2Future9 = exports.RtcmMessageMsm1Future9 = exports.RtcmMessageMsm7Future8 = exports.RtcmMessageMsm6Future8 = exports.RtcmMessageMsm5Future8 = exports.RtcmMessageMsm4Future8 = exports.RtcmMessageMsm3Future8 = void 0;
const rtcm_1 = require("../../../rtcm");
const common_1 = require("./common");
let RtcmMessageMsm1Gps = class RtcmMessageMsm1Gps extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Gps = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageMsm1Gps);
exports.RtcmMessageMsm1Gps = RtcmMessageMsm1Gps;
let RtcmMessageMsm2Gps = class RtcmMessageMsm2Gps extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Gps = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageMsm2Gps);
exports.RtcmMessageMsm2Gps = RtcmMessageMsm2Gps;
let RtcmMessageMsm3Gps = class RtcmMessageMsm3Gps extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Gps = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageMsm3Gps);
exports.RtcmMessageMsm3Gps = RtcmMessageMsm3Gps;
let RtcmMessageMsm4Gps = class RtcmMessageMsm4Gps extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Gps = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageMsm4Gps);
exports.RtcmMessageMsm4Gps = RtcmMessageMsm4Gps;
let RtcmMessageMsm5Gps = class RtcmMessageMsm5Gps extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Gps = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageMsm5Gps);
exports.RtcmMessageMsm5Gps = RtcmMessageMsm5Gps;
let RtcmMessageMsm6Gps = class RtcmMessageMsm6Gps extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Gps = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageMsm6Gps);
exports.RtcmMessageMsm6Gps = RtcmMessageMsm6Gps;
let RtcmMessageMsm7Gps = class RtcmMessageMsm7Gps extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Gps = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageMsm7Gps);
exports.RtcmMessageMsm7Gps = RtcmMessageMsm7Gps;
let RtcmMessageMsm1Glonass = class RtcmMessageMsm1Glonass extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Glonass = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageMsm1Glonass);
exports.RtcmMessageMsm1Glonass = RtcmMessageMsm1Glonass;
let RtcmMessageMsm2Glonass = class RtcmMessageMsm2Glonass extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Glonass = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageMsm2Glonass);
exports.RtcmMessageMsm2Glonass = RtcmMessageMsm2Glonass;
let RtcmMessageMsm3Glonass = class RtcmMessageMsm3Glonass extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Glonass = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageMsm3Glonass);
exports.RtcmMessageMsm3Glonass = RtcmMessageMsm3Glonass;
let RtcmMessageMsm4Glonass = class RtcmMessageMsm4Glonass extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Glonass = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageMsm4Glonass);
exports.RtcmMessageMsm4Glonass = RtcmMessageMsm4Glonass;
let RtcmMessageMsm5Glonass = class RtcmMessageMsm5Glonass extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Glonass = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageMsm5Glonass);
exports.RtcmMessageMsm5Glonass = RtcmMessageMsm5Glonass;
let RtcmMessageMsm6Glonass = class RtcmMessageMsm6Glonass extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Glonass = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageMsm6Glonass);
exports.RtcmMessageMsm6Glonass = RtcmMessageMsm6Glonass;
let RtcmMessageMsm7Glonass = class RtcmMessageMsm7Glonass extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Glonass = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageMsm7Glonass);
exports.RtcmMessageMsm7Glonass = RtcmMessageMsm7Glonass;
let RtcmMessageMsm1Galileo = class RtcmMessageMsm1Galileo extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Galileo = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GALILEO_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GALILEO)
], RtcmMessageMsm1Galileo);
exports.RtcmMessageMsm1Galileo = RtcmMessageMsm1Galileo;
let RtcmMessageMsm2Galileo = class RtcmMessageMsm2Galileo extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Galileo = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GALILEO_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GALILEO)
], RtcmMessageMsm2Galileo);
exports.RtcmMessageMsm2Galileo = RtcmMessageMsm2Galileo;
let RtcmMessageMsm3Galileo = class RtcmMessageMsm3Galileo extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Galileo = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GALILEO_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GALILEO)
], RtcmMessageMsm3Galileo);
exports.RtcmMessageMsm3Galileo = RtcmMessageMsm3Galileo;
let RtcmMessageMsm4Galileo = class RtcmMessageMsm4Galileo extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Galileo = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GALILEO_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GALILEO)
], RtcmMessageMsm4Galileo);
exports.RtcmMessageMsm4Galileo = RtcmMessageMsm4Galileo;
let RtcmMessageMsm5Galileo = class RtcmMessageMsm5Galileo extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Galileo = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GALILEO_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GALILEO)
], RtcmMessageMsm5Galileo);
exports.RtcmMessageMsm5Galileo = RtcmMessageMsm5Galileo;
let RtcmMessageMsm6Galileo = class RtcmMessageMsm6Galileo extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Galileo = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GALILEO_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GALILEO)
], RtcmMessageMsm6Galileo);
exports.RtcmMessageMsm6Galileo = RtcmMessageMsm6Galileo;
let RtcmMessageMsm7Galileo = class RtcmMessageMsm7Galileo extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Galileo = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GALILEO_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GALILEO)
], RtcmMessageMsm7Galileo);
exports.RtcmMessageMsm7Galileo = RtcmMessageMsm7Galileo;
let RtcmMessageMsm1Sbas = class RtcmMessageMsm1Sbas extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Sbas = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SBAS_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.SBAS)
], RtcmMessageMsm1Sbas);
exports.RtcmMessageMsm1Sbas = RtcmMessageMsm1Sbas;
let RtcmMessageMsm2Sbas = class RtcmMessageMsm2Sbas extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Sbas = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SBAS_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.SBAS)
], RtcmMessageMsm2Sbas);
exports.RtcmMessageMsm2Sbas = RtcmMessageMsm2Sbas;
let RtcmMessageMsm3Sbas = class RtcmMessageMsm3Sbas extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Sbas = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SBAS_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.SBAS)
], RtcmMessageMsm3Sbas);
exports.RtcmMessageMsm3Sbas = RtcmMessageMsm3Sbas;
let RtcmMessageMsm4Sbas = class RtcmMessageMsm4Sbas extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Sbas = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SBAS_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.SBAS)
], RtcmMessageMsm4Sbas);
exports.RtcmMessageMsm4Sbas = RtcmMessageMsm4Sbas;
let RtcmMessageMsm5Sbas = class RtcmMessageMsm5Sbas extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Sbas = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SBAS_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.SBAS)
], RtcmMessageMsm5Sbas);
exports.RtcmMessageMsm5Sbas = RtcmMessageMsm5Sbas;
let RtcmMessageMsm6Sbas = class RtcmMessageMsm6Sbas extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Sbas = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SBAS_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.SBAS)
], RtcmMessageMsm6Sbas);
exports.RtcmMessageMsm6Sbas = RtcmMessageMsm6Sbas;
let RtcmMessageMsm7Sbas = class RtcmMessageMsm7Sbas extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Sbas = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SBAS_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.SBAS)
], RtcmMessageMsm7Sbas);
exports.RtcmMessageMsm7Sbas = RtcmMessageMsm7Sbas;
let RtcmMessageMsm1Qzss = class RtcmMessageMsm1Qzss extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Qzss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.QZSS_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.QZSS)
], RtcmMessageMsm1Qzss);
exports.RtcmMessageMsm1Qzss = RtcmMessageMsm1Qzss;
let RtcmMessageMsm2Qzss = class RtcmMessageMsm2Qzss extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Qzss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.QZSS_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.QZSS)
], RtcmMessageMsm2Qzss);
exports.RtcmMessageMsm2Qzss = RtcmMessageMsm2Qzss;
let RtcmMessageMsm3Qzss = class RtcmMessageMsm3Qzss extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Qzss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.QZSS_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.QZSS)
], RtcmMessageMsm3Qzss);
exports.RtcmMessageMsm3Qzss = RtcmMessageMsm3Qzss;
let RtcmMessageMsm4Qzss = class RtcmMessageMsm4Qzss extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Qzss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.QZSS_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.QZSS)
], RtcmMessageMsm4Qzss);
exports.RtcmMessageMsm4Qzss = RtcmMessageMsm4Qzss;
let RtcmMessageMsm5Qzss = class RtcmMessageMsm5Qzss extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Qzss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.QZSS_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.QZSS)
], RtcmMessageMsm5Qzss);
exports.RtcmMessageMsm5Qzss = RtcmMessageMsm5Qzss;
let RtcmMessageMsm6Qzss = class RtcmMessageMsm6Qzss extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Qzss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.QZSS_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.QZSS)
], RtcmMessageMsm6Qzss);
exports.RtcmMessageMsm6Qzss = RtcmMessageMsm6Qzss;
let RtcmMessageMsm7Qzss = class RtcmMessageMsm7Qzss extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Qzss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.QZSS_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.QZSS)
], RtcmMessageMsm7Qzss);
exports.RtcmMessageMsm7Qzss = RtcmMessageMsm7Qzss;
let RtcmMessageMsm1Bds = class RtcmMessageMsm1Bds extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Bds = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.BEIDOU_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.BEIDOU)
], RtcmMessageMsm1Bds);
exports.RtcmMessageMsm1Bds = RtcmMessageMsm1Bds;
let RtcmMessageMsm2Bds = class RtcmMessageMsm2Bds extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Bds = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.BEIDOU_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.BEIDOU)
], RtcmMessageMsm2Bds);
exports.RtcmMessageMsm2Bds = RtcmMessageMsm2Bds;
let RtcmMessageMsm3Bds = class RtcmMessageMsm3Bds extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Bds = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.BEIDOU_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.BEIDOU)
], RtcmMessageMsm3Bds);
exports.RtcmMessageMsm3Bds = RtcmMessageMsm3Bds;
let RtcmMessageMsm4Bds = class RtcmMessageMsm4Bds extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Bds = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.BEIDOU_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.BEIDOU)
], RtcmMessageMsm4Bds);
exports.RtcmMessageMsm4Bds = RtcmMessageMsm4Bds;
let RtcmMessageMsm5Bds = class RtcmMessageMsm5Bds extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Bds = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.BEIDOU_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.BEIDOU)
], RtcmMessageMsm5Bds);
exports.RtcmMessageMsm5Bds = RtcmMessageMsm5Bds;
let RtcmMessageMsm6Bds = class RtcmMessageMsm6Bds extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Bds = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.BEIDOU_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.BEIDOU)
], RtcmMessageMsm6Bds);
exports.RtcmMessageMsm6Bds = RtcmMessageMsm6Bds;
let RtcmMessageMsm7Bds = class RtcmMessageMsm7Bds extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Bds = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.BEIDOU_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.BEIDOU)
], RtcmMessageMsm7Bds);
exports.RtcmMessageMsm7Bds = RtcmMessageMsm7Bds;
let RtcmMessageMsm1Irnss = class RtcmMessageMsm1Irnss extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Irnss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.IRNSS_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.IRNSS)
], RtcmMessageMsm1Irnss);
exports.RtcmMessageMsm1Irnss = RtcmMessageMsm1Irnss;
let RtcmMessageMsm2Irnss = class RtcmMessageMsm2Irnss extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Irnss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.IRNSS_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.IRNSS)
], RtcmMessageMsm2Irnss);
exports.RtcmMessageMsm2Irnss = RtcmMessageMsm2Irnss;
let RtcmMessageMsm3Irnss = class RtcmMessageMsm3Irnss extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Irnss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.IRNSS_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.IRNSS)
], RtcmMessageMsm3Irnss);
exports.RtcmMessageMsm3Irnss = RtcmMessageMsm3Irnss;
let RtcmMessageMsm4Irnss = class RtcmMessageMsm4Irnss extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Irnss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.IRNSS_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.IRNSS)
], RtcmMessageMsm4Irnss);
exports.RtcmMessageMsm4Irnss = RtcmMessageMsm4Irnss;
let RtcmMessageMsm5Irnss = class RtcmMessageMsm5Irnss extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Irnss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.IRNSS_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.IRNSS)
], RtcmMessageMsm5Irnss);
exports.RtcmMessageMsm5Irnss = RtcmMessageMsm5Irnss;
let RtcmMessageMsm6Irnss = class RtcmMessageMsm6Irnss extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Irnss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.IRNSS_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.IRNSS)
], RtcmMessageMsm6Irnss);
exports.RtcmMessageMsm6Irnss = RtcmMessageMsm6Irnss;
let RtcmMessageMsm7Irnss = class RtcmMessageMsm7Irnss extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Irnss = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.IRNSS_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.IRNSS)
], RtcmMessageMsm7Irnss);
exports.RtcmMessageMsm7Irnss = RtcmMessageMsm7Irnss;
let RtcmMessageMsm1Future1 = class RtcmMessageMsm1Future1 extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Future1 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_1_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm1Future1);
exports.RtcmMessageMsm1Future1 = RtcmMessageMsm1Future1;
let RtcmMessageMsm2Future1 = class RtcmMessageMsm2Future1 extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Future1 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_1_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm2Future1);
exports.RtcmMessageMsm2Future1 = RtcmMessageMsm2Future1;
let RtcmMessageMsm3Future1 = class RtcmMessageMsm3Future1 extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Future1 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_1_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm3Future1);
exports.RtcmMessageMsm3Future1 = RtcmMessageMsm3Future1;
let RtcmMessageMsm4Future1 = class RtcmMessageMsm4Future1 extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Future1 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_1_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm4Future1);
exports.RtcmMessageMsm4Future1 = RtcmMessageMsm4Future1;
let RtcmMessageMsm5Future1 = class RtcmMessageMsm5Future1 extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Future1 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_1_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm5Future1);
exports.RtcmMessageMsm5Future1 = RtcmMessageMsm5Future1;
let RtcmMessageMsm6Future1 = class RtcmMessageMsm6Future1 extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Future1 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_1_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm6Future1);
exports.RtcmMessageMsm6Future1 = RtcmMessageMsm6Future1;
let RtcmMessageMsm7Future1 = class RtcmMessageMsm7Future1 extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Future1 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_1_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm7Future1);
exports.RtcmMessageMsm7Future1 = RtcmMessageMsm7Future1;
let RtcmMessageMsm1Future2 = class RtcmMessageMsm1Future2 extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Future2 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_2_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm1Future2);
exports.RtcmMessageMsm1Future2 = RtcmMessageMsm1Future2;
let RtcmMessageMsm2Future2 = class RtcmMessageMsm2Future2 extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Future2 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_2_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm2Future2);
exports.RtcmMessageMsm2Future2 = RtcmMessageMsm2Future2;
let RtcmMessageMsm3Future2 = class RtcmMessageMsm3Future2 extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Future2 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_2_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm3Future2);
exports.RtcmMessageMsm3Future2 = RtcmMessageMsm3Future2;
let RtcmMessageMsm4Future2 = class RtcmMessageMsm4Future2 extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Future2 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_2_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm4Future2);
exports.RtcmMessageMsm4Future2 = RtcmMessageMsm4Future2;
let RtcmMessageMsm5Future2 = class RtcmMessageMsm5Future2 extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Future2 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_2_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm5Future2);
exports.RtcmMessageMsm5Future2 = RtcmMessageMsm5Future2;
let RtcmMessageMsm6Future2 = class RtcmMessageMsm6Future2 extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Future2 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_2_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm6Future2);
exports.RtcmMessageMsm6Future2 = RtcmMessageMsm6Future2;
let RtcmMessageMsm7Future2 = class RtcmMessageMsm7Future2 extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Future2 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_2_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm7Future2);
exports.RtcmMessageMsm7Future2 = RtcmMessageMsm7Future2;
let RtcmMessageMsm1Future3 = class RtcmMessageMsm1Future3 extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Future3 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_3_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm1Future3);
exports.RtcmMessageMsm1Future3 = RtcmMessageMsm1Future3;
let RtcmMessageMsm2Future3 = class RtcmMessageMsm2Future3 extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Future3 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_3_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm2Future3);
exports.RtcmMessageMsm2Future3 = RtcmMessageMsm2Future3;
let RtcmMessageMsm3Future3 = class RtcmMessageMsm3Future3 extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Future3 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_3_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm3Future3);
exports.RtcmMessageMsm3Future3 = RtcmMessageMsm3Future3;
let RtcmMessageMsm4Future3 = class RtcmMessageMsm4Future3 extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Future3 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_3_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm4Future3);
exports.RtcmMessageMsm4Future3 = RtcmMessageMsm4Future3;
let RtcmMessageMsm5Future3 = class RtcmMessageMsm5Future3 extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Future3 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_3_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm5Future3);
exports.RtcmMessageMsm5Future3 = RtcmMessageMsm5Future3;
let RtcmMessageMsm6Future3 = class RtcmMessageMsm6Future3 extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Future3 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_3_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm6Future3);
exports.RtcmMessageMsm6Future3 = RtcmMessageMsm6Future3;
let RtcmMessageMsm7Future3 = class RtcmMessageMsm7Future3 extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Future3 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_3_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm7Future3);
exports.RtcmMessageMsm7Future3 = RtcmMessageMsm7Future3;
let RtcmMessageMsm1Future4 = class RtcmMessageMsm1Future4 extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Future4 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_4_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm1Future4);
exports.RtcmMessageMsm1Future4 = RtcmMessageMsm1Future4;
let RtcmMessageMsm2Future4 = class RtcmMessageMsm2Future4 extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Future4 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_4_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm2Future4);
exports.RtcmMessageMsm2Future4 = RtcmMessageMsm2Future4;
let RtcmMessageMsm3Future4 = class RtcmMessageMsm3Future4 extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Future4 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_4_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm3Future4);
exports.RtcmMessageMsm3Future4 = RtcmMessageMsm3Future4;
let RtcmMessageMsm4Future4 = class RtcmMessageMsm4Future4 extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Future4 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_4_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm4Future4);
exports.RtcmMessageMsm4Future4 = RtcmMessageMsm4Future4;
let RtcmMessageMsm5Future4 = class RtcmMessageMsm5Future4 extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Future4 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_4_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm5Future4);
exports.RtcmMessageMsm5Future4 = RtcmMessageMsm5Future4;
let RtcmMessageMsm6Future4 = class RtcmMessageMsm6Future4 extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Future4 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_4_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm6Future4);
exports.RtcmMessageMsm6Future4 = RtcmMessageMsm6Future4;
let RtcmMessageMsm7Future4 = class RtcmMessageMsm7Future4 extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Future4 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_4_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm7Future4);
exports.RtcmMessageMsm7Future4 = RtcmMessageMsm7Future4;
let RtcmMessageMsm1Future5 = class RtcmMessageMsm1Future5 extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Future5 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_5_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm1Future5);
exports.RtcmMessageMsm1Future5 = RtcmMessageMsm1Future5;
let RtcmMessageMsm2Future5 = class RtcmMessageMsm2Future5 extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Future5 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_5_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm2Future5);
exports.RtcmMessageMsm2Future5 = RtcmMessageMsm2Future5;
let RtcmMessageMsm3Future5 = class RtcmMessageMsm3Future5 extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Future5 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_5_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm3Future5);
exports.RtcmMessageMsm3Future5 = RtcmMessageMsm3Future5;
let RtcmMessageMsm4Future5 = class RtcmMessageMsm4Future5 extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Future5 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_5_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm4Future5);
exports.RtcmMessageMsm4Future5 = RtcmMessageMsm4Future5;
let RtcmMessageMsm5Future5 = class RtcmMessageMsm5Future5 extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Future5 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_5_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm5Future5);
exports.RtcmMessageMsm5Future5 = RtcmMessageMsm5Future5;
let RtcmMessageMsm6Future5 = class RtcmMessageMsm6Future5 extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Future5 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_5_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm6Future5);
exports.RtcmMessageMsm6Future5 = RtcmMessageMsm6Future5;
let RtcmMessageMsm7Future5 = class RtcmMessageMsm7Future5 extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Future5 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_5_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm7Future5);
exports.RtcmMessageMsm7Future5 = RtcmMessageMsm7Future5;
let RtcmMessageMsm1Future6 = class RtcmMessageMsm1Future6 extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Future6 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_6_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm1Future6);
exports.RtcmMessageMsm1Future6 = RtcmMessageMsm1Future6;
let RtcmMessageMsm2Future6 = class RtcmMessageMsm2Future6 extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Future6 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_6_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm2Future6);
exports.RtcmMessageMsm2Future6 = RtcmMessageMsm2Future6;
let RtcmMessageMsm3Future6 = class RtcmMessageMsm3Future6 extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Future6 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_6_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm3Future6);
exports.RtcmMessageMsm3Future6 = RtcmMessageMsm3Future6;
let RtcmMessageMsm4Future6 = class RtcmMessageMsm4Future6 extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Future6 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_6_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm4Future6);
exports.RtcmMessageMsm4Future6 = RtcmMessageMsm4Future6;
let RtcmMessageMsm5Future6 = class RtcmMessageMsm5Future6 extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Future6 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_6_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm5Future6);
exports.RtcmMessageMsm5Future6 = RtcmMessageMsm5Future6;
let RtcmMessageMsm6Future6 = class RtcmMessageMsm6Future6 extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Future6 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_6_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm6Future6);
exports.RtcmMessageMsm6Future6 = RtcmMessageMsm6Future6;
let RtcmMessageMsm7Future6 = class RtcmMessageMsm7Future6 extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Future6 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_6_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm7Future6);
exports.RtcmMessageMsm7Future6 = RtcmMessageMsm7Future6;
let RtcmMessageMsm1Future7 = class RtcmMessageMsm1Future7 extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Future7 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_7_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm1Future7);
exports.RtcmMessageMsm1Future7 = RtcmMessageMsm1Future7;
let RtcmMessageMsm2Future7 = class RtcmMessageMsm2Future7 extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Future7 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_7_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm2Future7);
exports.RtcmMessageMsm2Future7 = RtcmMessageMsm2Future7;
let RtcmMessageMsm3Future7 = class RtcmMessageMsm3Future7 extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Future7 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_7_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm3Future7);
exports.RtcmMessageMsm3Future7 = RtcmMessageMsm3Future7;
let RtcmMessageMsm4Future7 = class RtcmMessageMsm4Future7 extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Future7 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_7_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm4Future7);
exports.RtcmMessageMsm4Future7 = RtcmMessageMsm4Future7;
let RtcmMessageMsm5Future7 = class RtcmMessageMsm5Future7 extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Future7 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_7_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm5Future7);
exports.RtcmMessageMsm5Future7 = RtcmMessageMsm5Future7;
let RtcmMessageMsm6Future7 = class RtcmMessageMsm6Future7 extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Future7 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_7_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm6Future7);
exports.RtcmMessageMsm6Future7 = RtcmMessageMsm6Future7;
let RtcmMessageMsm7Future7 = class RtcmMessageMsm7Future7 extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Future7 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_7_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm7Future7);
exports.RtcmMessageMsm7Future7 = RtcmMessageMsm7Future7;
let RtcmMessageMsm1Future8 = class RtcmMessageMsm1Future8 extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Future8 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_8_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm1Future8);
exports.RtcmMessageMsm1Future8 = RtcmMessageMsm1Future8;
let RtcmMessageMsm2Future8 = class RtcmMessageMsm2Future8 extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Future8 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_8_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm2Future8);
exports.RtcmMessageMsm2Future8 = RtcmMessageMsm2Future8;
let RtcmMessageMsm3Future8 = class RtcmMessageMsm3Future8 extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Future8 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_8_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm3Future8);
exports.RtcmMessageMsm3Future8 = RtcmMessageMsm3Future8;
let RtcmMessageMsm4Future8 = class RtcmMessageMsm4Future8 extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Future8 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_8_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm4Future8);
exports.RtcmMessageMsm4Future8 = RtcmMessageMsm4Future8;
let RtcmMessageMsm5Future8 = class RtcmMessageMsm5Future8 extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Future8 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_8_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm5Future8);
exports.RtcmMessageMsm5Future8 = RtcmMessageMsm5Future8;
let RtcmMessageMsm6Future8 = class RtcmMessageMsm6Future8 extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Future8 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_8_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm6Future8);
exports.RtcmMessageMsm6Future8 = RtcmMessageMsm6Future8;
let RtcmMessageMsm7Future8 = class RtcmMessageMsm7Future8 extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Future8 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_8_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm7Future8);
exports.RtcmMessageMsm7Future8 = RtcmMessageMsm7Future8;
let RtcmMessageMsm1Future9 = class RtcmMessageMsm1Future9 extends common_1.RtcmMessageMsm1 {
};
RtcmMessageMsm1Future9 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_9_MSM1),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm1Future9);
exports.RtcmMessageMsm1Future9 = RtcmMessageMsm1Future9;
let RtcmMessageMsm2Future9 = class RtcmMessageMsm2Future9 extends common_1.RtcmMessageMsm2 {
};
RtcmMessageMsm2Future9 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_9_MSM2),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm2Future9);
exports.RtcmMessageMsm2Future9 = RtcmMessageMsm2Future9;
let RtcmMessageMsm3Future9 = class RtcmMessageMsm3Future9 extends common_1.RtcmMessageMsm3 {
};
RtcmMessageMsm3Future9 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_9_MSM3),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm3Future9);
exports.RtcmMessageMsm3Future9 = RtcmMessageMsm3Future9;
let RtcmMessageMsm4Future9 = class RtcmMessageMsm4Future9 extends common_1.RtcmMessageMsm4 {
};
RtcmMessageMsm4Future9 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_9_MSM4),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm4Future9);
exports.RtcmMessageMsm4Future9 = RtcmMessageMsm4Future9;
let RtcmMessageMsm5Future9 = class RtcmMessageMsm5Future9 extends common_1.RtcmMessageMsm5 {
};
RtcmMessageMsm5Future9 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_9_MSM5),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm5Future9);
exports.RtcmMessageMsm5Future9 = RtcmMessageMsm5Future9;
let RtcmMessageMsm6Future9 = class RtcmMessageMsm6Future9 extends common_1.RtcmMessageMsm6 {
};
RtcmMessageMsm6Future9 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_9_MSM6),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm6Future9);
exports.RtcmMessageMsm6Future9 = RtcmMessageMsm6Future9;
let RtcmMessageMsm7Future9 = class RtcmMessageMsm7Future9 extends common_1.RtcmMessageMsm7 {
};
RtcmMessageMsm7Future9 = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.FUTURE_9_MSM7),
    rtcm_1.Since(rtcm_1.RtcmVersion.FUTURE),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.FUTURE)
], RtcmMessageMsm7Future9);
exports.RtcmMessageMsm7Future9 = RtcmMessageMsm7Future9;
exports.referenceSignals = {
    [rtcm_1.RtcmNavSystem.GPS]: {
        L1: [1575.42, 'L1C'],
        L2: [12227.6, 'L2P'],
        L3: [1176.45, 'L5I']
    },
    [rtcm_1.RtcmNavSystem.GLONASS]: {
        G1: [(k) => 1602 + k * 9 / 16, 'L1C'],
        G2: [(k) => 1246 + k * 7 / 16, 'L2C']
    },
    [rtcm_1.RtcmNavSystem.GALILEO]: {
        E1: [1575.42, 'L1B'],
        E5A: [1176.45, 'L5I'],
        E5B: [1207.14, 'L7I'],
        E5AB: [1191.795, 'L8I'],
        E6: [1278.75, 'L6B'],
    },
    [rtcm_1.RtcmNavSystem.SBAS]: {
        L1: [1575.42, 'L1C'],
        L5: [1176.45, 'L5I']
    },
    [rtcm_1.RtcmNavSystem.QZSS]: {
        L1: [1575.42, 'L1C'],
        L2: [1227.60, 'L2S'],
        L5: [1176.45, 'L5I']
    },
    [rtcm_1.RtcmNavSystem.BEIDOU]: {
        B1: [1561.098, 'L2I'],
        B2: [1207.140, 'L7I'],
        B3: [1268.52, 'L6I']
    },
    [rtcm_1.RtcmNavSystem.IRNSS]: {
        L5: [1176.45, 'L5A']
    },
    [rtcm_1.RtcmNavSystem.FUTURE]: {}
};
exports.signalIdMapping = {
    [rtcm_1.RtcmNavSystem.GPS]: {
        2: [1, 'L1', '1C'],
        3: [1, 'L1', '1P'],
        4: [1, 'L1', '1W'],
        8: [2, 'L2', '2C'],
        9: [2, 'L2', '2P'],
        10: [2, 'L2', '2W'],
        15: [2, 'L2', '2S'],
        16: [2, 'L2', '2L'],
        17: [2, 'L2', '2X'],
        22: [5, 'L5', '5I'],
        23: [5, 'L5', '5Q'],
        24: [5, 'L5', '5X'],
        30: [1, 'L1', '1S'],
        31: [1, 'L1', '1L'],
        32: [1, 'L1', '1X']
    },
    [rtcm_1.RtcmNavSystem.GLONASS]: {
        1: [1, 'G1', '1C'],
        2: [1, 'G1', '1P'],
        8: [2, 'G2', '2C'],
        9: [2, 'G2', '2P']
    },
    [rtcm_1.RtcmNavSystem.GALILEO]: {
        2: [1, 'E1', '1C'],
        3: [1, 'E1', '1A'],
        4: [1, 'E1', '1B'],
        5: [1, 'E1', '1X'],
        6: [1, 'E1', '1Z'],
        8: [6, 'E6', '6C'],
        9: [6, 'E6', '6A'],
        10: [6, 'E6', '6B'],
        11: [6, 'E6', '6X'],
        12: [6, 'E6', '6Z'],
        14: [5, 'E5B', '7I'],
        15: [5, 'E5B', '7Q'],
        16: [5, 'E5B', '7X'],
        18: [5, 'E5AB', '8I'],
        19: [5, 'E5AB', '8Q'],
        20: [5, 'E5AB', '8X'],
        22: [5, 'E5A', '5I'],
        23: [5, 'E5A', '5Q'],
        24: [5, 'E5A', '5X']
    },
    [rtcm_1.RtcmNavSystem.SBAS]: {
        2: [1, 'L1', '1C'],
        22: [5, 'L5', '5I'],
        23: [5, 'L5', '5Q'],
        24: [5, 'L5', '5X']
    },
    [rtcm_1.RtcmNavSystem.QZSS]: {
        2: [1, 'L1', '1C'],
        9: [0, 'LEX', '6S'],
        10: [0, 'LEX', '6L'],
        11: [0, 'LEX', '6X'],
        15: [2, 'L2', '2S'],
        16: [2, 'L2', '2L'],
        17: [2, 'L2', '2X'],
        22: [5, 'L5', '5I'],
        23: [5, 'L5', '5Q'],
        24: [5, 'L5', '5X'],
        30: [1, 'L1', '1S'],
        31: [1, 'L1', '1L'],
        32: [1, 'L1', '1X']
    },
    [rtcm_1.RtcmNavSystem.BEIDOU]: {
        2: [1, 'B1', '2I'],
        3: [1, 'B1', '2Q'],
        4: [1, 'B1', '2X'],
        8: [3, 'B3', '6I'],
        9: [3, 'B3', '6Q'],
        10: [3, 'B3', '6X'],
        14: [2, 'B2', '7I'],
        15: [2, 'B2', '7Q'],
        16: [2, 'B2', '7X']
    },
    [rtcm_1.RtcmNavSystem.IRNSS]: {
        22: [5, 'L5', '5A']
    },
    [rtcm_1.RtcmNavSystem.FUTURE]: {}
};
