// @gnss/rtcm/build/messages/observations/msm/satellite.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Msm7SatelliteData = exports.Msm5SatelliteData = exports.Msm57SatelliteData = exports.Msm6SatelliteData = exports.Msm4SatelliteData = exports.Msm46SatelliteData = exports.Msm4567SatelliteData = exports.Msm3SatelliteData = exports.Msm2SatelliteData = exports.Msm1SatelliteData = exports.Msm123SatelliteData = exports.MsmSatelliteData = void 0;
const utils_1 = require("../../../utils");
const decode_encode_1 = require("../../../decode-encode");
const constants_1 = require("../../../constants");
class MsmSatelliteData {
    constructor(_internalGuard) { }
}
exports.MsmSatelliteData = MsmSatelliteData;
MsmSatelliteData.construct = utils_1.constructWithPropertiesInternalGuardPlus;
class Msm123SatelliteData extends MsmSatelliteData {
    getRoughRangeMicroseconds(roughRangeIntegerMilliseconds) {
        return roughRangeIntegerMilliseconds * 1000
            + this.roughRangeModulo1Millisecond * 1000 / 2 ** 10;
    }
    set roughRangeMicroseconds(range) { this.roughRangeModulo1Millisecond = Math.round((range % 1000) * 2 ** 10 / 1000); }
    getRoughRangeMeters(roughRangeIntegerMilliseconds) {
        return this.getRoughRangeMicroseconds(roughRangeIntegerMilliseconds) * constants_1.SPEED_OF_LIGHT / 1000000;
    }
    set roughRangeMeters(range) { this.roughRangeMicroseconds = range * 1000000 / constants_1.SPEED_OF_LIGHT; }
}
__decorate([
    decode_encode_1.DF(398),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], Msm123SatelliteData.prototype, "roughRangeModulo1Millisecond", void 0);
exports.Msm123SatelliteData = Msm123SatelliteData;
class Msm1SatelliteData extends Msm123SatelliteData {
}
exports.Msm1SatelliteData = Msm1SatelliteData;
class Msm2SatelliteData extends Msm123SatelliteData {
}
exports.Msm2SatelliteData = Msm2SatelliteData;
class Msm3SatelliteData extends Msm123SatelliteData {
}
exports.Msm3SatelliteData = Msm3SatelliteData;
class Msm4567SatelliteData extends MsmSatelliteData {
    get roughRangeMicroseconds() {
        return this.roughRangeIntegerMilliseconds * 1000 + this.roughRangeModulo1Millisecond * 1000 / 2 ** 10;
    }
    set roughRangeMicroseconds(range) {
        this.roughRangeIntegerMilliseconds = Math.floor(range / 1000);
        this.roughRangeModulo1Millisecond = Math.round((range % 1000) * (2 ** 10) / 1000);
    }
    get roughRangeMeters() { return this.roughRangeMicroseconds * constants_1.SPEED_OF_LIGHT / 1000000; }
    set roughRangeMeters(range) { this.roughRangeMicroseconds = range * 1000000 / constants_1.SPEED_OF_LIGHT; }
}
__decorate([
    decode_encode_1.DF(397),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], Msm4567SatelliteData.prototype, "roughRangeIntegerMilliseconds", void 0);
__decorate([
    decode_encode_1.UInt(4),
    decode_encode_1.IfInstanceOf(() => Msm57SatelliteData),
    __metadata("design:type", Number)
], Msm4567SatelliteData.prototype, "extendedInformation", void 0);
__decorate([
    decode_encode_1.DF(398),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], Msm4567SatelliteData.prototype, "roughRangeModulo1Millisecond", void 0);
__decorate([
    decode_encode_1.DF(399),
    decode_encode_1.Int(14),
    decode_encode_1.IfInstanceOf(() => Msm57SatelliteData),
    __metadata("design:type", Number)
], Msm4567SatelliteData.prototype, "roughPhaserangeRateMetersPerSecond", void 0);
exports.Msm4567SatelliteData = Msm4567SatelliteData;
class Msm46SatelliteData extends Msm4567SatelliteData {
}
exports.Msm46SatelliteData = Msm46SatelliteData;
class Msm4SatelliteData extends Msm46SatelliteData {
}
exports.Msm4SatelliteData = Msm4SatelliteData;
class Msm6SatelliteData extends Msm46SatelliteData {
}
exports.Msm6SatelliteData = Msm6SatelliteData;
class Msm57SatelliteData extends Msm4567SatelliteData {
}
exports.Msm57SatelliteData = Msm57SatelliteData;
class Msm5SatelliteData extends Msm57SatelliteData {
}
exports.Msm5SatelliteData = Msm5SatelliteData;
class Msm7SatelliteData extends Msm57SatelliteData {
}
exports.Msm7SatelliteData = Msm7SatelliteData;
