// @ntrip/caster/build/index.js
"use strict";
/*
 * This file is part of the @ntrip/caster distribution (https://github.com/node-ntrip/caster).
 * Copyright (c) 2020 Nebojsa Cvetkovic.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleLogger = exports.Sourcetable = exports.AutoSourceEntry = exports.Mountpoint = exports.Client = exports.Server = exports.Connection = exports.Caster = exports.NtripHTTPParser = void 0;
const ntrip_http_parser_1 = require("./util/ntrip-http-parser");
var ntrip_http_parser_2 = require("./util/ntrip-http-parser");
Object.defineProperty(exports, "NtripHTTPParser", { enumerable: true, get: function () { return ntrip_http_parser_2.NtripHTTPParser; } });
ntrip_http_parser_1.NtripHTTPParser.bind();
var caster_1 = require("./caster");
Object.defineProperty(exports, "Caster", { enumerable: true, get: function () { return caster_1.Caster; } });
var connection_1 = require("./connection");
Object.defineProperty(exports, "Connection", { enumerable: true, get: function () { return connection_1.Connection; } });
Object.defineProperty(exports, "Server", { enumerable: true, get: function () { return connection_1.Server; } });
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return connection_1.Client; } });
var mountpoint_1 = require("./mountpoint");
Object.defineProperty(exports, "Mountpoint", { enumerable: true, get: function () { return mountpoint_1.Mountpoint; } });
var auto_source_entry_1 = require("./util/auto-source-entry");
Object.defineProperty(exports, "AutoSourceEntry", { enumerable: true, get: function () { return auto_source_entry_1.AutoSourceEntry; } });
var sourcetable_1 = require("./sourcetable");
Object.defineProperty(exports, "Sourcetable", { enumerable: true, get: function () { return sourcetable_1.Sourcetable; } });
__exportStar(require("./transport"), exports);
var simple_logger_1 = require("./util/simple-logger");
Object.defineProperty(exports, "SimpleLogger", { enumerable: true, get: function () { return simple_logger_1.SimpleLogger; } });
