// @gnss/rtcm/build/messages/state-space/glonass.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmMessageSsrGlonassHighRateClockCorrection = exports.GlonassHighRateClockCorrectionSatelliteData = exports.RtcmMessageSsrGlonassUra = exports.GlonassUraSatelliteData = exports.RtcmMessageSsrGlonassCodeBias = exports.GlonassCodeBiasSatelliteData = exports.GlonassCodeBiasSatelliteCodeBias = exports.RtcmMessageSsrGlonassCombinedCorrection = exports.GlonassCombinedCorrectionSatelliteData = exports.RtcmMessageSsrGlonassClockCorrection = exports.GlonassClockCorrectionSatelliteData = exports.RtcmMessageSsrGlonassOrbitCorrection = exports.GlonassOrbitCorrectionSatelliteData = exports.RtcmMessageSsrGlonassCorrection = void 0;
const decode_encode_1 = require("../../decode-encode");
const rtcm_1 = require("../../rtcm");
const utils_1 = require("../../utils");
let RtcmMessageSsrGlonassCorrection = class RtcmMessageSsrGlonassCorrection extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(385),
    decode_encode_1.UInt(17),
    __metadata("design:type", Number)
], RtcmMessageSsrGlonassCorrection.prototype, "epochTime", void 0);
__decorate([
    decode_encode_1.DF(391),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageSsrGlonassCorrection.prototype, "ssrUpdateInterval", void 0);
__decorate([
    decode_encode_1.DF(388),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageSsrGlonassCorrection.prototype, "multipleMessageIndicator", void 0);
__decorate([
    decode_encode_1.DF(375),
    decode_encode_1.Bool,
    decode_encode_1.IfInstanceOf(() => RtcmMessageSsrGlonassOrbitCorrection),
    __metadata("design:type", Boolean)
], RtcmMessageSsrGlonassCorrection.prototype, "satelliteReferenceDatum", void 0);
__decorate([
    decode_encode_1.DF(413),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageSsrGlonassCorrection.prototype, "iodSsr", void 0);
__decorate([
    decode_encode_1.DF(414),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageSsrGlonassCorrection.prototype, "ssrProviderId", void 0);
__decorate([
    decode_encode_1.DF(415),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageSsrGlonassCorrection.prototype, "ssrSolutionId", void 0);
__decorate([
    decode_encode_1.DF(387),
    decode_encode_1.ArrLength(6, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageSsrGlonassCorrection.prototype, "_satelliteCount", void 0);
RtcmMessageSsrGlonassCorrection = __decorate([
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageSsrGlonassCorrection);
exports.RtcmMessageSsrGlonassCorrection = RtcmMessageSsrGlonassCorrection;
class GlonassOrbitCorrectionSatelliteData {
}
__decorate([
    decode_encode_1.DF(384),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], GlonassOrbitCorrectionSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(392),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GlonassOrbitCorrectionSatelliteData.prototype, "iod", void 0);
__decorate([
    decode_encode_1.DF(365),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], GlonassOrbitCorrectionSatelliteData.prototype, "deltaRadial", void 0);
__decorate([
    decode_encode_1.DF(366),
    decode_encode_1.Int(20),
    __metadata("design:type", Number)
], GlonassOrbitCorrectionSatelliteData.prototype, "deltaAlongTrack", void 0);
__decorate([
    decode_encode_1.DF(367),
    decode_encode_1.Int(20),
    __metadata("design:type", Number)
], GlonassOrbitCorrectionSatelliteData.prototype, "deltaCrossTrack", void 0);
__decorate([
    decode_encode_1.DF(368),
    decode_encode_1.Int(21),
    __metadata("design:type", Number)
], GlonassOrbitCorrectionSatelliteData.prototype, "dotDeltaRadial", void 0);
__decorate([
    decode_encode_1.DF(369),
    decode_encode_1.Int(19),
    __metadata("design:type", Number)
], GlonassOrbitCorrectionSatelliteData.prototype, "dotDeltaAlongTrack", void 0);
__decorate([
    decode_encode_1.DF(370),
    decode_encode_1.Int(19),
    __metadata("design:type", Number)
], GlonassOrbitCorrectionSatelliteData.prototype, "dotDeltaCrossTrack", void 0);
exports.GlonassOrbitCorrectionSatelliteData = GlonassOrbitCorrectionSatelliteData;
let RtcmMessageSsrGlonassOrbitCorrection = class RtcmMessageSsrGlonassOrbitCorrection extends RtcmMessageSsrGlonassCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassOrbitCorrectionSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGlonassOrbitCorrection.prototype, "satellites", void 0);
RtcmMessageSsrGlonassOrbitCorrection = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GLONASS_ORBIT_CORRECTION)
], RtcmMessageSsrGlonassOrbitCorrection);
exports.RtcmMessageSsrGlonassOrbitCorrection = RtcmMessageSsrGlonassOrbitCorrection;
class GlonassClockCorrectionSatelliteData {
}
__decorate([
    decode_encode_1.DF(384),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], GlonassClockCorrectionSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(376),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], GlonassClockCorrectionSatelliteData.prototype, "deltaClockC0", void 0);
__decorate([
    decode_encode_1.DF(377),
    decode_encode_1.Int(21),
    __metadata("design:type", Number)
], GlonassClockCorrectionSatelliteData.prototype, "deltaClockC1", void 0);
__decorate([
    decode_encode_1.DF(378),
    decode_encode_1.Int(27),
    __metadata("design:type", Number)
], GlonassClockCorrectionSatelliteData.prototype, "deltaClockC2", void 0);
exports.GlonassClockCorrectionSatelliteData = GlonassClockCorrectionSatelliteData;
let RtcmMessageSsrGlonassClockCorrection = class RtcmMessageSsrGlonassClockCorrection extends RtcmMessageSsrGlonassCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassClockCorrectionSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGlonassClockCorrection.prototype, "satellites", void 0);
RtcmMessageSsrGlonassClockCorrection = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GLONASS_CLOCK_CORRECTION)
], RtcmMessageSsrGlonassClockCorrection);
exports.RtcmMessageSsrGlonassClockCorrection = RtcmMessageSsrGlonassClockCorrection;
class GlonassCombinedCorrectionSatelliteData extends GlonassOrbitCorrectionSatelliteData {
}
__decorate([
    decode_encode_1.DF(376),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], GlonassCombinedCorrectionSatelliteData.prototype, "deltaClockC0", void 0);
__decorate([
    decode_encode_1.DF(377),
    decode_encode_1.Int(21),
    __metadata("design:type", Number)
], GlonassCombinedCorrectionSatelliteData.prototype, "deltaClockC1", void 0);
__decorate([
    decode_encode_1.DF(378),
    decode_encode_1.Int(27),
    __metadata("design:type", Number)
], GlonassCombinedCorrectionSatelliteData.prototype, "deltaClockC2", void 0);
exports.GlonassCombinedCorrectionSatelliteData = GlonassCombinedCorrectionSatelliteData;
let RtcmMessageSsrGlonassCombinedCorrection = class RtcmMessageSsrGlonassCombinedCorrection extends RtcmMessageSsrGlonassCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassCombinedCorrectionSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGlonassCombinedCorrection.prototype, "satellites", void 0);
RtcmMessageSsrGlonassCombinedCorrection = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GLONASS_COMBINED_CORRECTION)
], RtcmMessageSsrGlonassCombinedCorrection);
exports.RtcmMessageSsrGlonassCombinedCorrection = RtcmMessageSsrGlonassCombinedCorrection;
class GlonassCodeBiasSatelliteCodeBias {
}
__decorate([
    decode_encode_1.DF(381),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], GlonassCodeBiasSatelliteCodeBias.prototype, "signalAndTrackingModeIndicator", void 0);
__decorate([
    decode_encode_1.DF(383),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], GlonassCodeBiasSatelliteCodeBias.prototype, "codeBias", void 0);
exports.GlonassCodeBiasSatelliteCodeBias = GlonassCodeBiasSatelliteCodeBias;
class GlonassCodeBiasSatelliteData {
    constructor(internalGuard) { }
}
GlonassCodeBiasSatelliteData.construct = utils_1.constructWithPropertiesInternalGuard;
__decorate([
    decode_encode_1.DF(384),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], GlonassCodeBiasSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(379),
    decode_encode_1.ArrLength(5, 'codeBiases'),
    __metadata("design:type", Number)
], GlonassCodeBiasSatelliteData.prototype, "_codeBiasCount", void 0);
__decorate([
    decode_encode_1.Arr('_codeBiasCount', decode_encode_1.Obj(GlonassCodeBiasSatelliteCodeBias)),
    __metadata("design:type", Array)
], GlonassCodeBiasSatelliteData.prototype, "codeBiases", void 0);
exports.GlonassCodeBiasSatelliteData = GlonassCodeBiasSatelliteData;
let RtcmMessageSsrGlonassCodeBias = class RtcmMessageSsrGlonassCodeBias extends RtcmMessageSsrGlonassCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassCodeBiasSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGlonassCodeBias.prototype, "satellites", void 0);
RtcmMessageSsrGlonassCodeBias = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GLONASS_CODE_BIAS)
], RtcmMessageSsrGlonassCodeBias);
exports.RtcmMessageSsrGlonassCodeBias = RtcmMessageSsrGlonassCodeBias;
class GlonassUraSatelliteData {
}
__decorate([
    decode_encode_1.DF(384),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], GlonassUraSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(389),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GlonassUraSatelliteData.prototype, "ssrUra", void 0);
exports.GlonassUraSatelliteData = GlonassUraSatelliteData;
let RtcmMessageSsrGlonassUra = class RtcmMessageSsrGlonassUra extends RtcmMessageSsrGlonassCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassUraSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGlonassUra.prototype, "satellites", void 0);
RtcmMessageSsrGlonassUra = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GLONASS_URA)
], RtcmMessageSsrGlonassUra);
exports.RtcmMessageSsrGlonassUra = RtcmMessageSsrGlonassUra;
class GlonassHighRateClockCorrectionSatelliteData {
}
__decorate([
    decode_encode_1.DF(384),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], GlonassHighRateClockCorrectionSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(390),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], GlonassHighRateClockCorrectionSatelliteData.prototype, "highRateClockCorrection", void 0);
exports.GlonassHighRateClockCorrectionSatelliteData = GlonassHighRateClockCorrectionSatelliteData;
let RtcmMessageSsrGlonassHighRateClockCorrection = class RtcmMessageSsrGlonassHighRateClockCorrection extends RtcmMessageSsrGlonassCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GlonassHighRateClockCorrectionSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGlonassHighRateClockCorrection.prototype, "satellites", void 0);
RtcmMessageSsrGlonassHighRateClockCorrection = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GLONASS_HIGH_RATE_CLOCK_CORRECTION)
], RtcmMessageSsrGlonassHighRateClockCorrection);
exports.RtcmMessageSsrGlonassHighRateClockCorrection = RtcmMessageSsrGlonassHighRateClockCorrection;
