// @gnss/rtcm/build/messages/state-space/gps.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmMessageSsrGpsHighRateClockCorrection = exports.GpsHighRateClockCorrectionSatelliteData = exports.RtcmMessageSsrGpsUra = exports.GpsUraSatelliteData = exports.RtcmMessageSsrGpsCodeBias = exports.GpsCodeBiasSatelliteData = exports.GpsCodeBiasSatelliteCodeBias = exports.RtcmMessageSsrGpsCombinedCorrection = exports.GpsCombinedCorrectionSatelliteData = exports.RtcmMessageSsrGpsClockCorrection = exports.GpsClockCorrectionSatelliteData = exports.RtcmMessageSsrGpsOrbitCorrection = exports.GpsOrbitCorrectionSatelliteData = exports.RtcmMessageSsrGpsCorrection = void 0;
const decode_encode_1 = require("../../decode-encode");
const rtcm_1 = require("../../rtcm");
const utils_1 = require("../../utils");
let RtcmMessageSsrGpsCorrection = class RtcmMessageSsrGpsCorrection extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(385),
    decode_encode_1.UInt(20),
    __metadata("design:type", Number)
], RtcmMessageSsrGpsCorrection.prototype, "epochTime", void 0);
__decorate([
    decode_encode_1.DF(391),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageSsrGpsCorrection.prototype, "ssrUpdateInterval", void 0);
__decorate([
    decode_encode_1.DF(388),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageSsrGpsCorrection.prototype, "multipleMessageIndicator", void 0);
__decorate([
    decode_encode_1.DF(375),
    decode_encode_1.Bool,
    decode_encode_1.IfInstanceOf(() => RtcmMessageSsrGpsOrbitCorrection),
    __metadata("design:type", Boolean)
], RtcmMessageSsrGpsCorrection.prototype, "satelliteReferenceDatum", void 0);
__decorate([
    decode_encode_1.DF(413),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageSsrGpsCorrection.prototype, "iodSsr", void 0);
__decorate([
    decode_encode_1.DF(414),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageSsrGpsCorrection.prototype, "ssrProviderId", void 0);
__decorate([
    decode_encode_1.DF(415),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageSsrGpsCorrection.prototype, "ssrSolutionId", void 0);
__decorate([
    decode_encode_1.DF(387),
    decode_encode_1.ArrLength(6, 'satellites'),
    __metadata("design:type", Number)
], RtcmMessageSsrGpsCorrection.prototype, "_satelliteCount", void 0);
RtcmMessageSsrGpsCorrection = __decorate([
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageSsrGpsCorrection);
exports.RtcmMessageSsrGpsCorrection = RtcmMessageSsrGpsCorrection;
class GpsOrbitCorrectionSatelliteData {
}
__decorate([
    decode_encode_1.DF(68),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsOrbitCorrectionSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(71),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], GpsOrbitCorrectionSatelliteData.prototype, "iode", void 0);
__decorate([
    decode_encode_1.DF(365),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], GpsOrbitCorrectionSatelliteData.prototype, "deltaRadial", void 0);
__decorate([
    decode_encode_1.DF(366),
    decode_encode_1.Int(20),
    __metadata("design:type", Number)
], GpsOrbitCorrectionSatelliteData.prototype, "deltaAlongTrack", void 0);
__decorate([
    decode_encode_1.DF(367),
    decode_encode_1.Int(20),
    __metadata("design:type", Number)
], GpsOrbitCorrectionSatelliteData.prototype, "deltaCrossTrack", void 0);
__decorate([
    decode_encode_1.DF(368),
    decode_encode_1.Int(21),
    __metadata("design:type", Number)
], GpsOrbitCorrectionSatelliteData.prototype, "dotDeltaRadial", void 0);
__decorate([
    decode_encode_1.DF(369),
    decode_encode_1.Int(19),
    __metadata("design:type", Number)
], GpsOrbitCorrectionSatelliteData.prototype, "dotDeltaAlongTrack", void 0);
__decorate([
    decode_encode_1.DF(370),
    decode_encode_1.Int(19),
    __metadata("design:type", Number)
], GpsOrbitCorrectionSatelliteData.prototype, "dotDeltaCrossTrack", void 0);
exports.GpsOrbitCorrectionSatelliteData = GpsOrbitCorrectionSatelliteData;
let RtcmMessageSsrGpsOrbitCorrection = class RtcmMessageSsrGpsOrbitCorrection extends RtcmMessageSsrGpsCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsOrbitCorrectionSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGpsOrbitCorrection.prototype, "satellites", void 0);
RtcmMessageSsrGpsOrbitCorrection = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GPS_ORBIT_CORRECTION)
], RtcmMessageSsrGpsOrbitCorrection);
exports.RtcmMessageSsrGpsOrbitCorrection = RtcmMessageSsrGpsOrbitCorrection;
class GpsClockCorrectionSatelliteData {
}
__decorate([
    decode_encode_1.DF(68),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsClockCorrectionSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(376),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], GpsClockCorrectionSatelliteData.prototype, "deltaClockC0", void 0);
__decorate([
    decode_encode_1.DF(377),
    decode_encode_1.Int(21),
    __metadata("design:type", Number)
], GpsClockCorrectionSatelliteData.prototype, "deltaClockC1", void 0);
__decorate([
    decode_encode_1.DF(378),
    decode_encode_1.Int(27),
    __metadata("design:type", Number)
], GpsClockCorrectionSatelliteData.prototype, "deltaClockC2", void 0);
exports.GpsClockCorrectionSatelliteData = GpsClockCorrectionSatelliteData;
let RtcmMessageSsrGpsClockCorrection = class RtcmMessageSsrGpsClockCorrection extends RtcmMessageSsrGpsCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsClockCorrectionSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGpsClockCorrection.prototype, "satellites", void 0);
RtcmMessageSsrGpsClockCorrection = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GPS_CLOCK_CORRECTION)
], RtcmMessageSsrGpsClockCorrection);
exports.RtcmMessageSsrGpsClockCorrection = RtcmMessageSsrGpsClockCorrection;
class GpsCombinedCorrectionSatelliteData extends GpsOrbitCorrectionSatelliteData {
}
__decorate([
    decode_encode_1.DF(376),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], GpsCombinedCorrectionSatelliteData.prototype, "deltaClockC0", void 0);
__decorate([
    decode_encode_1.DF(377),
    decode_encode_1.Int(21),
    __metadata("design:type", Number)
], GpsCombinedCorrectionSatelliteData.prototype, "deltaClockC1", void 0);
__decorate([
    decode_encode_1.DF(378),
    decode_encode_1.Int(27),
    __metadata("design:type", Number)
], GpsCombinedCorrectionSatelliteData.prototype, "deltaClockC2", void 0);
exports.GpsCombinedCorrectionSatelliteData = GpsCombinedCorrectionSatelliteData;
let RtcmMessageSsrGpsCombinedCorrection = class RtcmMessageSsrGpsCombinedCorrection extends RtcmMessageSsrGpsCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsCombinedCorrectionSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGpsCombinedCorrection.prototype, "satellites", void 0);
RtcmMessageSsrGpsCombinedCorrection = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GPS_COMBINED_CORRECTION)
], RtcmMessageSsrGpsCombinedCorrection);
exports.RtcmMessageSsrGpsCombinedCorrection = RtcmMessageSsrGpsCombinedCorrection;
class GpsCodeBiasSatelliteCodeBias {
}
__decorate([
    decode_encode_1.DF(380),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], GpsCodeBiasSatelliteCodeBias.prototype, "signalAndTrackingModeIndicator", void 0);
__decorate([
    decode_encode_1.DF(383),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], GpsCodeBiasSatelliteCodeBias.prototype, "codeBias", void 0);
exports.GpsCodeBiasSatelliteCodeBias = GpsCodeBiasSatelliteCodeBias;
class GpsCodeBiasSatelliteData {
}
GpsCodeBiasSatelliteData.construct = utils_1.constructWithProperties;
__decorate([
    decode_encode_1.DF(68),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsCodeBiasSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(379),
    decode_encode_1.ArrLength(5, 'codeBiases'),
    __metadata("design:type", Number)
], GpsCodeBiasSatelliteData.prototype, "_codeBiasCount", void 0);
__decorate([
    decode_encode_1.Arr('_codeBiasCount', decode_encode_1.Obj(GpsCodeBiasSatelliteCodeBias)),
    __metadata("design:type", Array)
], GpsCodeBiasSatelliteData.prototype, "codeBiases", void 0);
exports.GpsCodeBiasSatelliteData = GpsCodeBiasSatelliteData;
let RtcmMessageSsrGpsCodeBias = class RtcmMessageSsrGpsCodeBias extends RtcmMessageSsrGpsCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsCodeBiasSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGpsCodeBias.prototype, "satellites", void 0);
RtcmMessageSsrGpsCodeBias = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GPS_CODE_BIAS)
], RtcmMessageSsrGpsCodeBias);
exports.RtcmMessageSsrGpsCodeBias = RtcmMessageSsrGpsCodeBias;
class GpsUraSatelliteData {
}
__decorate([
    decode_encode_1.DF(68),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsUraSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(389),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsUraSatelliteData.prototype, "ssrUra", void 0);
exports.GpsUraSatelliteData = GpsUraSatelliteData;
let RtcmMessageSsrGpsUra = class RtcmMessageSsrGpsUra extends RtcmMessageSsrGpsCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsUraSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGpsUra.prototype, "satellites", void 0);
RtcmMessageSsrGpsUra = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GPS_URA)
], RtcmMessageSsrGpsUra);
exports.RtcmMessageSsrGpsUra = RtcmMessageSsrGpsUra;
class GpsHighRateClockCorrectionSatelliteData {
}
__decorate([
    decode_encode_1.DF(68),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], GpsHighRateClockCorrectionSatelliteData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(390),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], GpsHighRateClockCorrectionSatelliteData.prototype, "highRateClockCorrection", void 0);
exports.GpsHighRateClockCorrectionSatelliteData = GpsHighRateClockCorrectionSatelliteData;
let RtcmMessageSsrGpsHighRateClockCorrection = class RtcmMessageSsrGpsHighRateClockCorrection extends RtcmMessageSsrGpsCorrection {
};
__decorate([
    decode_encode_1.Arr('_satelliteCount', decode_encode_1.Obj(GpsHighRateClockCorrectionSatelliteData)),
    __metadata("design:type", Array)
], RtcmMessageSsrGpsHighRateClockCorrection.prototype, "satellites", void 0);
RtcmMessageSsrGpsHighRateClockCorrection = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.SSR_GPS_HIGH_RATE_CLOCK_CORRECTION)
], RtcmMessageSsrGpsHighRateClockCorrection);
exports.RtcmMessageSsrGpsHighRateClockCorrection = RtcmMessageSsrGpsHighRateClockCorrection;
