// @gnss/rtcm/build/messages/transformation.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmMessageResidualsPlaneGridRepresentation = exports.RtcmMessageResidualsEllipsoidalGridRepresentation = exports.RtcmMessageResidualsGridRepresentation = exports.ResidualsPlaneGridRepresentationData = exports.ResidualsEllipsoidalGridRepresentationData = exports.RtcmMessageMolodenskiBadekasTransformationParameters = exports.RtcmMessageHelmertAbridgedMolodenskiTransformationParameters = exports.RtcmMessageBaseTransformationParameters = void 0;
const decode_encode_1 = require("../decode-encode");
const rtcm_1 = require("../rtcm");
class RtcmMessageBaseTransformationParameters extends rtcm_1.RtcmMessage {
}
__decorate([
    decode_encode_1.DF(144),
    decode_encode_1.Latin1(5),
    __metadata("design:type", String)
], RtcmMessageBaseTransformationParameters.prototype, "sourceName", void 0);
__decorate([
    decode_encode_1.DF(146),
    decode_encode_1.Latin1(5),
    __metadata("design:type", String)
], RtcmMessageBaseTransformationParameters.prototype, "targetName", void 0);
__decorate([
    decode_encode_1.DF(147),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "systemIdentificationNumber", void 0);
__decorate([
    decode_encode_1.DF(148),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "utilizedTransformationMessageIndicator", void 0);
__decorate([
    decode_encode_1.DF(149),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "plateNumber", void 0);
__decorate([
    decode_encode_1.DF(150),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "computationIndicator", void 0);
__decorate([
    decode_encode_1.DF(151),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "heightIndicator", void 0);
__decorate([
    decode_encode_1.DF(152),
    decode_encode_1.Int(19),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "originLatitude", void 0);
__decorate([
    decode_encode_1.DF(153),
    decode_encode_1.Int(20),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "originLongitude", void 0);
__decorate([
    decode_encode_1.DF(154),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "nsExtension", void 0);
__decorate([
    decode_encode_1.DF(155),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "ewExtension", void 0);
__decorate([
    decode_encode_1.DF(156),
    decode_encode_1.Int(23),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "xTranslation", void 0);
__decorate([
    decode_encode_1.DF(157),
    decode_encode_1.Int(23),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "yTranslation", void 0);
__decorate([
    decode_encode_1.DF(158),
    decode_encode_1.Int(23),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "zTranslation", void 0);
__decorate([
    decode_encode_1.DF(159),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "xRotation", void 0);
__decorate([
    decode_encode_1.DF(160),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "yRotation", void 0);
__decorate([
    decode_encode_1.DF(161),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "zRotation", void 0);
__decorate([
    decode_encode_1.DF(162),
    decode_encode_1.Int(25),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "scaleCorrection", void 0);
__decorate([
    decode_encode_1.DF(163),
    decode_encode_1.Int(35),
    decode_encode_1.IfInstanceOf(() => RtcmMessageMolodenskiBadekasTransformationParameters),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "mbRotationPointXCoordinate", void 0);
__decorate([
    decode_encode_1.DF(164),
    decode_encode_1.Int(35),
    decode_encode_1.IfInstanceOf(() => RtcmMessageMolodenskiBadekasTransformationParameters),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "mbRotationPointYCoordinate", void 0);
__decorate([
    decode_encode_1.DF(165),
    decode_encode_1.Int(35),
    decode_encode_1.IfInstanceOf(() => RtcmMessageMolodenskiBadekasTransformationParameters),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "mbRotationPointZCoordinate", void 0);
__decorate([
    decode_encode_1.DF(166),
    decode_encode_1.UInt(24),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "sourceSystemEllipsoidSemiMajorAxis", void 0);
__decorate([
    decode_encode_1.DF(167),
    decode_encode_1.UInt(25),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "sourceSystemEllipsoidSemiMinorAxis", void 0);
__decorate([
    decode_encode_1.DF(168),
    decode_encode_1.UInt(24),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "targetSystemEllipsoidSemiMajorAxis", void 0);
__decorate([
    decode_encode_1.DF(169),
    decode_encode_1.UInt(25),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "targetSystemEllipsoidSemiMinorAxis", void 0);
__decorate([
    decode_encode_1.DF(214),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "horizontalHelmertMolodenskiQualityIndicator", void 0);
__decorate([
    decode_encode_1.DF(215),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], RtcmMessageBaseTransformationParameters.prototype, "verticalHelmertMolodenskiQualityIndicator", void 0);
exports.RtcmMessageBaseTransformationParameters = RtcmMessageBaseTransformationParameters;
let RtcmMessageHelmertAbridgedMolodenskiTransformationParameters = class RtcmMessageHelmertAbridgedMolodenskiTransformationParameters extends RtcmMessageBaseTransformationParameters {
};
RtcmMessageHelmertAbridgedMolodenskiTransformationParameters = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.HELMERT_ABRIDGED_MOLODENSKI_TRANSFORMATION_PARAMETERS),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1)
], RtcmMessageHelmertAbridgedMolodenskiTransformationParameters);
exports.RtcmMessageHelmertAbridgedMolodenskiTransformationParameters = RtcmMessageHelmertAbridgedMolodenskiTransformationParameters;
let RtcmMessageMolodenskiBadekasTransformationParameters = class RtcmMessageMolodenskiBadekasTransformationParameters extends RtcmMessageBaseTransformationParameters {
};
RtcmMessageMolodenskiBadekasTransformationParameters = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.MOLODENSKI_BADEKAS_TRANSFORMATION_PARAMETERS),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1)
], RtcmMessageMolodenskiBadekasTransformationParameters);
exports.RtcmMessageMolodenskiBadekasTransformationParameters = RtcmMessageMolodenskiBadekasTransformationParameters;
class ResidualsEllipsoidalGridRepresentationData {
}
__decorate([
    decode_encode_1.DF(199),
    decode_encode_1.Int(9),
    __metadata("design:type", Number)
], ResidualsEllipsoidalGridRepresentationData.prototype, "latitudeResidual", void 0);
__decorate([
    decode_encode_1.DF(200),
    decode_encode_1.Int(9),
    __metadata("design:type", Number)
], ResidualsEllipsoidalGridRepresentationData.prototype, "longitudeResidual", void 0);
__decorate([
    decode_encode_1.DF(201),
    decode_encode_1.Int(9),
    __metadata("design:type", Number)
], ResidualsEllipsoidalGridRepresentationData.prototype, "heightResidual", void 0);
exports.ResidualsEllipsoidalGridRepresentationData = ResidualsEllipsoidalGridRepresentationData;
class ResidualsPlaneGridRepresentationData {
}
__decorate([
    decode_encode_1.DF(209),
    decode_encode_1.Int(9),
    __metadata("design:type", Number)
], ResidualsPlaneGridRepresentationData.prototype, "localNorthingResidual", void 0);
__decorate([
    decode_encode_1.DF(210),
    decode_encode_1.Int(9),
    __metadata("design:type", Number)
], ResidualsPlaneGridRepresentationData.prototype, "localEastingResidual", void 0);
__decorate([
    decode_encode_1.DF(211),
    decode_encode_1.Int(9),
    __metadata("design:type", Number)
], ResidualsPlaneGridRepresentationData.prototype, "localHeightResidual", void 0);
exports.ResidualsPlaneGridRepresentationData = ResidualsPlaneGridRepresentationData;
class RtcmMessageResidualsGridRepresentation extends rtcm_1.RtcmMessage {
}
__decorate([
    decode_encode_1.DF(147),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], RtcmMessageResidualsGridRepresentation.prototype, "systemIdentificationNumber", void 0);
__decorate([
    decode_encode_1.DF(190),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageResidualsGridRepresentation.prototype, "horizontalShiftIndicator", void 0);
__decorate([
    decode_encode_1.DF(191),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageResidualsGridRepresentation.prototype, "verticalShiftIndicator", void 0);
exports.RtcmMessageResidualsGridRepresentation = RtcmMessageResidualsGridRepresentation;
let RtcmMessageResidualsEllipsoidalGridRepresentation = class RtcmMessageResidualsEllipsoidalGridRepresentation extends RtcmMessageResidualsGridRepresentation {
};
__decorate([
    decode_encode_1.DF(192),
    decode_encode_1.Int(21),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "latitudeGridsOrigin", void 0);
__decorate([
    decode_encode_1.DF(193),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "longitudeGridsOrigin", void 0);
__decorate([
    decode_encode_1.DF(194),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "nsGridAreaExtension", void 0);
__decorate([
    decode_encode_1.DF(195),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "ewGridAreaExtension", void 0);
__decorate([
    decode_encode_1.DF(196),
    decode_encode_1.Int(8),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "meanLatitudeOffset", void 0);
__decorate([
    decode_encode_1.DF(197),
    decode_encode_1.Int(8),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "meanLongitudeOffset", void 0);
__decorate([
    decode_encode_1.DF(198),
    decode_encode_1.Int(15),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "meanHeightOffset", void 0);
__decorate([
    decode_encode_1.Arr(16, decode_encode_1.Obj(ResidualsEllipsoidalGridRepresentationData)),
    __metadata("design:type", Array)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "residuals", void 0);
__decorate([
    decode_encode_1.DF(212),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "horizontalInterpolationMethodIndicator", void 0);
__decorate([
    decode_encode_1.DF(213),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "verticalInterpolationMethodIndicator", void 0);
__decorate([
    decode_encode_1.DF(216),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "horizontalGridQualityIndicator", void 0);
__decorate([
    decode_encode_1.DF(217),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "verticalGridQualityIndicator", void 0);
__decorate([
    decode_encode_1.DF(51),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageResidualsEllipsoidalGridRepresentation.prototype, "modifiedJulianDay", void 0);
RtcmMessageResidualsEllipsoidalGridRepresentation = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.RESIDUALS_ELLIPSOIDAL_GRID_REPRESENTATION),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1)
], RtcmMessageResidualsEllipsoidalGridRepresentation);
exports.RtcmMessageResidualsEllipsoidalGridRepresentation = RtcmMessageResidualsEllipsoidalGridRepresentation;
let RtcmMessageResidualsPlaneGridRepresentation = class RtcmMessageResidualsPlaneGridRepresentation extends RtcmMessageResidualsGridRepresentation {
};
__decorate([
    decode_encode_1.DF(202),
    decode_encode_1.Int(25),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "northingOrigin", void 0);
__decorate([
    decode_encode_1.DF(203),
    decode_encode_1.Int(26),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "eastingOrigin", void 0);
__decorate([
    decode_encode_1.DF(204),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "nsGridAreaExtension", void 0);
__decorate([
    decode_encode_1.DF(205),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "ewGridAreaExtension", void 0);
__decorate([
    decode_encode_1.DF(206),
    decode_encode_1.Int(10),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "meanLocalNorthingResidual", void 0);
__decorate([
    decode_encode_1.DF(207),
    decode_encode_1.Int(10),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "meanLocalEastingResidual", void 0);
__decorate([
    decode_encode_1.DF(208),
    decode_encode_1.Int(15),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "meanLocalHeightResidual", void 0);
__decorate([
    decode_encode_1.Arr(16, decode_encode_1.Obj(ResidualsPlaneGridRepresentationData)),
    __metadata("design:type", Array)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "residuals", void 0);
__decorate([
    decode_encode_1.DF(212),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "horizontalInterpolationMethodIndicator", void 0);
__decorate([
    decode_encode_1.DF(213),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "verticalInterpolationMethodIndicator", void 0);
__decorate([
    decode_encode_1.DF(216),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "horizontalGridQualityIndicator", void 0);
__decorate([
    decode_encode_1.DF(217),
    decode_encode_1.UInt(3),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "verticalGridQualityIndicator", void 0);
__decorate([
    decode_encode_1.DF(51),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageResidualsPlaneGridRepresentation.prototype, "modifiedJulianDay", void 0);
RtcmMessageResidualsPlaneGridRepresentation = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.RESIDUALS_PLANE_GRID_REPRESENTATION),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1)
], RtcmMessageResidualsPlaneGridRepresentation);
exports.RtcmMessageResidualsPlaneGridRepresentation = RtcmMessageResidualsPlaneGridRepresentation;
