// @gnss/rtcm/build/messages/auxiliary/ephemeris.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcmMessageIrnssSatelliteEphemerisData = exports.RtcmMessageGalileoINavSatelliteEphemerisData = exports.RtcmMessageGalileoFNavSatelliteEphemerisData = exports.RtcmMessageQzssSatelliteEphemerisData = exports.RtcmMessageBdsSatelliteEphemerisData = exports.RtcmMessageGlonassSatelliteEphemerisData = exports.RtcmMessageGpsSatelliteEphemerisData = void 0;
const decode_encode_1 = require("../../decode-encode");
const rtcm_1 = require("../../rtcm");
let RtcmMessageGpsSatelliteEphemerisData = class RtcmMessageGpsSatelliteEphemerisData extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(9),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(76),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "weekNumber", void 0);
__decorate([
    decode_encode_1.DF(77),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "svAccuracy", void 0);
__decorate([
    decode_encode_1.DF(78),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "codeOnL2", void 0);
__decorate([
    decode_encode_1.DF(79),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "idot", void 0);
__decorate([
    decode_encode_1.DF(71),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "iode", void 0);
__decorate([
    decode_encode_1.DF(81),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "t_oc", void 0);
__decorate([
    decode_encode_1.DF(82),
    decode_encode_1.Int(8),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "a_f2", void 0);
__decorate([
    decode_encode_1.DF(83),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "a_f1", void 0);
__decorate([
    decode_encode_1.DF(84),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "a_f0", void 0);
__decorate([
    decode_encode_1.DF(85),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "iodc", void 0);
__decorate([
    decode_encode_1.DF(86),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "c_rs", void 0);
__decorate([
    decode_encode_1.DF(87),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "deltaN", void 0);
__decorate([
    decode_encode_1.DF(88),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "m_0", void 0);
__decorate([
    decode_encode_1.DF(89),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "c_uc", void 0);
__decorate([
    decode_encode_1.DF(90),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "e", void 0);
__decorate([
    decode_encode_1.DF(91),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "c_us", void 0);
__decorate([
    decode_encode_1.DF(92),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "a_1_2", void 0);
__decorate([
    decode_encode_1.DF(93),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "t_oe", void 0);
__decorate([
    decode_encode_1.DF(94),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "c_ic", void 0);
__decorate([
    decode_encode_1.DF(95),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "omega_0", void 0);
__decorate([
    decode_encode_1.DF(96),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "c_is", void 0);
__decorate([
    decode_encode_1.DF(97),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "i_0", void 0);
__decorate([
    decode_encode_1.DF(98),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "c_rc", void 0);
__decorate([
    decode_encode_1.DF(99),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "w", void 0);
__decorate([
    decode_encode_1.DF(100),
    decode_encode_1.Int(24),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "omegadot", void 0);
__decorate([
    decode_encode_1.DF(101),
    decode_encode_1.Int(8),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "t_gd", void 0);
__decorate([
    decode_encode_1.DF(102),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "svHealth", void 0);
__decorate([
    decode_encode_1.DF(103),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "l2PDataFlag", void 0);
__decorate([
    decode_encode_1.DF(137),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGpsSatelliteEphemerisData.prototype, "fitInterval", void 0);
RtcmMessageGpsSatelliteEphemerisData = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GPS_SATELLITE_EPHEMERIS_DATA),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GPS)
], RtcmMessageGpsSatelliteEphemerisData);
exports.RtcmMessageGpsSatelliteEphemerisData = RtcmMessageGpsSatelliteEphemerisData;
let RtcmMessageGlonassSatelliteEphemerisData = class RtcmMessageGlonassSatelliteEphemerisData extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(38),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(40),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "satelliteFrequencyChannelNumber", void 0);
__decorate([
    decode_encode_1.DF(104),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "almanacHealth", void 0);
__decorate([
    decode_encode_1.DF(105),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "almanacHealthAvailabilityIndicator", void 0);
__decorate([
    decode_encode_1.DF(106),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "p1", void 0);
__decorate([
    decode_encode_1.DF(107),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "t_k", void 0);
__decorate([
    decode_encode_1.DF(108),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "b_nMsbWord", void 0);
__decorate([
    decode_encode_1.DF(109),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "p2", void 0);
__decorate([
    decode_encode_1.DF(110),
    decode_encode_1.UInt(7),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "t_b", void 0);
__decorate([
    decode_encode_1.DF(111),
    decode_encode_1.IntSM(24),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "x_n_1", void 0);
__decorate([
    decode_encode_1.DF(112),
    decode_encode_1.IntSM(27),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "x_n", void 0);
__decorate([
    decode_encode_1.DF(113),
    decode_encode_1.IntSM(5),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "x_n_2", void 0);
__decorate([
    decode_encode_1.DF(114),
    decode_encode_1.IntSM(24),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "y_n_1", void 0);
__decorate([
    decode_encode_1.DF(115),
    decode_encode_1.IntSM(27),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "y_n", void 0);
__decorate([
    decode_encode_1.DF(116),
    decode_encode_1.IntSM(5),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "y_n_2", void 0);
__decorate([
    decode_encode_1.DF(117),
    decode_encode_1.IntSM(24),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "z_n_1", void 0);
__decorate([
    decode_encode_1.DF(118),
    decode_encode_1.IntSM(27),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "z_n", void 0);
__decorate([
    decode_encode_1.DF(119),
    decode_encode_1.IntSM(5),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "z_n_2", void 0);
__decorate([
    decode_encode_1.DF(120),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "p3", void 0);
__decorate([
    decode_encode_1.DF(121),
    decode_encode_1.IntSM(11),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "gamma_n", void 0);
__decorate([
    decode_encode_1.DF(122),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "mP", void 0);
__decorate([
    decode_encode_1.DF(123),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "mL_n_3", void 0);
__decorate([
    decode_encode_1.DF(124),
    decode_encode_1.IntSM(22),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "t_n", void 0);
__decorate([
    decode_encode_1.DF(125),
    decode_encode_1.IntSM(5),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "mDeltaT_n", void 0);
__decorate([
    decode_encode_1.DF(126),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "e_n", void 0);
__decorate([
    decode_encode_1.DF(127),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "mP4", void 0);
__decorate([
    decode_encode_1.DF(128),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "mF_t", void 0);
__decorate([
    decode_encode_1.DF(129),
    decode_encode_1.UInt(11),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "mN_t", void 0);
__decorate([
    decode_encode_1.DF(130),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "mM", void 0);
__decorate([
    decode_encode_1.DF(131),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "remainderValid", void 0);
__decorate([
    decode_encode_1.DF(132),
    decode_encode_1.UInt(11),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "n_a", void 0);
__decorate([
    decode_encode_1.DF(133),
    decode_encode_1.IntSM(32),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "t_c", void 0);
__decorate([
    decode_encode_1.DF(134),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "mN_4", void 0);
__decorate([
    decode_encode_1.DF(135),
    decode_encode_1.IntSM(22),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "t_gps", void 0);
__decorate([
    decode_encode_1.DF(136),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "mL_n_5", void 0);
__decorate([
    decode_encode_1.DF(1),
    decode_encode_1.Skip(7),
    __metadata("design:type", void 0)
], RtcmMessageGlonassSatelliteEphemerisData.prototype, "skip2", void 0);
RtcmMessageGlonassSatelliteEphemerisData = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GLONASS_SATELLITE_EPHEMERIS_DATA),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_1),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GLONASS)
], RtcmMessageGlonassSatelliteEphemerisData);
exports.RtcmMessageGlonassSatelliteEphemerisData = RtcmMessageGlonassSatelliteEphemerisData;
let RtcmMessageBdsSatelliteEphemerisData = class RtcmMessageBdsSatelliteEphemerisData extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(488),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(489),
    decode_encode_1.UInt(13),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "weekNumber", void 0);
__decorate([
    decode_encode_1.DF(490),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "svUrai", void 0);
__decorate([
    decode_encode_1.DF(491),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "idot", void 0);
__decorate([
    decode_encode_1.DF(492),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "aode", void 0);
__decorate([
    decode_encode_1.DF(493),
    decode_encode_1.UInt(17),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "t_oc", void 0);
__decorate([
    decode_encode_1.DF(494),
    decode_encode_1.Int(11),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "a_2", void 0);
__decorate([
    decode_encode_1.DF(495),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "a_1", void 0);
__decorate([
    decode_encode_1.DF(496),
    decode_encode_1.Int(24),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "a_0", void 0);
__decorate([
    decode_encode_1.DF(497),
    decode_encode_1.UInt(5),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "aodc", void 0);
__decorate([
    decode_encode_1.DF(498),
    decode_encode_1.Int(18),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "c_rs", void 0);
__decorate([
    decode_encode_1.DF(499),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "deltaN", void 0);
__decorate([
    decode_encode_1.DF(500),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "m_0", void 0);
__decorate([
    decode_encode_1.DF(501),
    decode_encode_1.Int(18),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "c_uc", void 0);
__decorate([
    decode_encode_1.DF(502),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "e", void 0);
__decorate([
    decode_encode_1.DF(503),
    decode_encode_1.Int(18),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "c_us", void 0);
__decorate([
    decode_encode_1.DF(504),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "a_1_2", void 0);
__decorate([
    decode_encode_1.DF(505),
    decode_encode_1.UInt(17),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "t_oe", void 0);
__decorate([
    decode_encode_1.DF(506),
    decode_encode_1.Int(18),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "c_ic", void 0);
__decorate([
    decode_encode_1.DF(507),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "omega_0", void 0);
__decorate([
    decode_encode_1.DF(508),
    decode_encode_1.Int(18),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "c_is", void 0);
__decorate([
    decode_encode_1.DF(509),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "i_0", void 0);
__decorate([
    decode_encode_1.DF(510),
    decode_encode_1.Int(18),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "c_rc", void 0);
__decorate([
    decode_encode_1.DF(511),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "w", void 0);
__decorate([
    decode_encode_1.DF(512),
    decode_encode_1.Int(24),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "omegaDot", void 0);
__decorate([
    decode_encode_1.DF(513),
    decode_encode_1.Int(10),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "t_gd1", void 0);
__decorate([
    decode_encode_1.DF(514),
    decode_encode_1.Int(10),
    __metadata("design:type", Number)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "t_gd2", void 0);
__decorate([
    decode_encode_1.DF(515),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageBdsSatelliteEphemerisData.prototype, "svHealth", void 0);
RtcmMessageBdsSatelliteEphemerisData = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.BEIDOU_SATELLITE_EPHEMERIS_DATA),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.BEIDOU)
], RtcmMessageBdsSatelliteEphemerisData);
exports.RtcmMessageBdsSatelliteEphemerisData = RtcmMessageBdsSatelliteEphemerisData;
let RtcmMessageQzssSatelliteEphemerisData = class RtcmMessageQzssSatelliteEphemerisData extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(429),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(430),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "t_oc", void 0);
__decorate([
    decode_encode_1.DF(431),
    decode_encode_1.Int(8),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "a_f2", void 0);
__decorate([
    decode_encode_1.DF(432),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "a_f1", void 0);
__decorate([
    decode_encode_1.DF(433),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "a_f0", void 0);
__decorate([
    decode_encode_1.DF(434),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "iode", void 0);
__decorate([
    decode_encode_1.DF(435),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "c_rs", void 0);
__decorate([
    decode_encode_1.DF(436),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "deltaN_0", void 0);
__decorate([
    decode_encode_1.DF(437),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "m_0", void 0);
__decorate([
    decode_encode_1.DF(438),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "c_uc", void 0);
__decorate([
    decode_encode_1.DF(439),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "e", void 0);
__decorate([
    decode_encode_1.DF(440),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "c_us", void 0);
__decorate([
    decode_encode_1.DF(441),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "a_1_2", void 0);
__decorate([
    decode_encode_1.DF(442),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "t_oe", void 0);
__decorate([
    decode_encode_1.DF(443),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "c_ic", void 0);
__decorate([
    decode_encode_1.DF(444),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "omega_0", void 0);
__decorate([
    decode_encode_1.DF(445),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "c_is", void 0);
__decorate([
    decode_encode_1.DF(446),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "i_0", void 0);
__decorate([
    decode_encode_1.DF(447),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "c_rc", void 0);
__decorate([
    decode_encode_1.DF(448),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "w_n", void 0);
__decorate([
    decode_encode_1.DF(449),
    decode_encode_1.Int(24),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "omegaDot", void 0);
__decorate([
    decode_encode_1.DF(450),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "i_0_dot", void 0);
__decorate([
    decode_encode_1.DF(451),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "l2ChannelCodes", void 0);
__decorate([
    decode_encode_1.DF(452),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "weekNumber", void 0);
__decorate([
    decode_encode_1.DF(453),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "ura", void 0);
__decorate([
    decode_encode_1.DF(454),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "svHealth", void 0);
__decorate([
    decode_encode_1.DF(455),
    decode_encode_1.Int(8),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "t_gd", void 0);
__decorate([
    decode_encode_1.DF(456),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "iodc", void 0);
__decorate([
    decode_encode_1.DF(457),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageQzssSatelliteEphemerisData.prototype, "fitInterval", void 0);
RtcmMessageQzssSatelliteEphemerisData = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.QZSS_SATELLITE_EPHEMERIS_DATA),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.QZSS)
], RtcmMessageQzssSatelliteEphemerisData);
exports.RtcmMessageQzssSatelliteEphemerisData = RtcmMessageQzssSatelliteEphemerisData;
let RtcmMessageGalileoFNavSatelliteEphemerisData = class RtcmMessageGalileoFNavSatelliteEphemerisData extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(252),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(289),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "weekNumber", void 0);
__decorate([
    decode_encode_1.DF(290),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "iodnav", void 0);
__decorate([
    decode_encode_1.DF(291),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "svSisAccuracy", void 0);
__decorate([
    decode_encode_1.DF(292),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "idot", void 0);
__decorate([
    decode_encode_1.DF(293),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "t_oc", void 0);
__decorate([
    decode_encode_1.DF(294),
    decode_encode_1.Int(6),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "a_f2", void 0);
__decorate([
    decode_encode_1.DF(295),
    decode_encode_1.Int(21),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "a_f1", void 0);
__decorate([
    decode_encode_1.DF(296),
    decode_encode_1.Int(31),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "a_f0", void 0);
__decorate([
    decode_encode_1.DF(297),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "c_rs", void 0);
__decorate([
    decode_encode_1.DF(298),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "deltaN", void 0);
__decorate([
    decode_encode_1.DF(299),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "m_0", void 0);
__decorate([
    decode_encode_1.DF(300),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "c_uc", void 0);
__decorate([
    decode_encode_1.DF(301),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "e", void 0);
__decorate([
    decode_encode_1.DF(302),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "c_us", void 0);
__decorate([
    decode_encode_1.DF(303),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "a_1_2", void 0);
__decorate([
    decode_encode_1.DF(304),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "t_oe", void 0);
__decorate([
    decode_encode_1.DF(305),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "c_ic", void 0);
__decorate([
    decode_encode_1.DF(306),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "omega_0", void 0);
__decorate([
    decode_encode_1.DF(307),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "c_is", void 0);
__decorate([
    decode_encode_1.DF(308),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "i_0", void 0);
__decorate([
    decode_encode_1.DF(309),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "c_rc", void 0);
__decorate([
    decode_encode_1.DF(310),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "w", void 0);
__decorate([
    decode_encode_1.DF(311),
    decode_encode_1.Int(24),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "omegaDot", void 0);
__decorate([
    decode_encode_1.DF(312),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "bgd_e5a_e1", void 0);
__decorate([
    decode_encode_1.DF(314),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "navSignalHealthStatus", void 0);
__decorate([
    decode_encode_1.DF(315),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "navDataValidityStatus", void 0);
__decorate([
    decode_encode_1.DF(1),
    decode_encode_1.Skip(7),
    __metadata("design:type", void 0)
], RtcmMessageGalileoFNavSatelliteEphemerisData.prototype, "skip2", void 0);
RtcmMessageGalileoFNavSatelliteEphemerisData = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GALILEO_F_NAV_SATELLITE_EPHEMERIS_DATA),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_2),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GALILEO)
], RtcmMessageGalileoFNavSatelliteEphemerisData);
exports.RtcmMessageGalileoFNavSatelliteEphemerisData = RtcmMessageGalileoFNavSatelliteEphemerisData;
let RtcmMessageGalileoINavSatelliteEphemerisData = class RtcmMessageGalileoINavSatelliteEphemerisData extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(252),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(289),
    decode_encode_1.UInt(12),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "weekNumber", void 0);
__decorate([
    decode_encode_1.DF(290),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "iodnav", void 0);
__decorate([
    decode_encode_1.DF(286),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "sisaIndex", void 0);
__decorate([
    decode_encode_1.DF(292),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "idot", void 0);
__decorate([
    decode_encode_1.DF(293),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "t_oc", void 0);
__decorate([
    decode_encode_1.DF(294),
    decode_encode_1.Int(6),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "a_f2", void 0);
__decorate([
    decode_encode_1.DF(295),
    decode_encode_1.Int(21),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "a_f1", void 0);
__decorate([
    decode_encode_1.DF(296),
    decode_encode_1.Int(31),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "a_f0", void 0);
__decorate([
    decode_encode_1.DF(297),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "c_rs", void 0);
__decorate([
    decode_encode_1.DF(298),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "deltaN", void 0);
__decorate([
    decode_encode_1.DF(299),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "m_0", void 0);
__decorate([
    decode_encode_1.DF(300),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "c_uc", void 0);
__decorate([
    decode_encode_1.DF(301),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "e", void 0);
__decorate([
    decode_encode_1.DF(302),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "c_us", void 0);
__decorate([
    decode_encode_1.DF(303),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "a_1_2", void 0);
__decorate([
    decode_encode_1.DF(304),
    decode_encode_1.UInt(14),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "t_oe", void 0);
__decorate([
    decode_encode_1.DF(305),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "c_ic", void 0);
__decorate([
    decode_encode_1.DF(306),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "omega_0", void 0);
__decorate([
    decode_encode_1.DF(307),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "c_is", void 0);
__decorate([
    decode_encode_1.DF(308),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "i_0", void 0);
__decorate([
    decode_encode_1.DF(309),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "c_rc", void 0);
__decorate([
    decode_encode_1.DF(310),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "w", void 0);
__decorate([
    decode_encode_1.DF(311),
    decode_encode_1.Int(24),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "omegaDot", void 0);
__decorate([
    decode_encode_1.DF(312),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "bgd_e5a_e1", void 0);
__decorate([
    decode_encode_1.DF(313),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "bgd_e5b_e1", void 0);
__decorate([
    decode_encode_1.DF(316),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "e5bSignalHealthStatus", void 0);
__decorate([
    decode_encode_1.DF(317),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "e5bDataValidityStatus", void 0);
__decorate([
    decode_encode_1.DF(287),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "e1bSignalHealthStatus", void 0);
__decorate([
    decode_encode_1.DF(288),
    decode_encode_1.UInt(1),
    __metadata("design:type", Number)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "e1bDataValidityStatus", void 0);
__decorate([
    decode_encode_1.DF(1),
    decode_encode_1.Skip(2),
    __metadata("design:type", void 0)
], RtcmMessageGalileoINavSatelliteEphemerisData.prototype, "skip2", void 0);
RtcmMessageGalileoINavSatelliteEphemerisData = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.GALILEO_I_NAV_SATELLITE_EPHEMERIS_DATA),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.GALILEO)
], RtcmMessageGalileoINavSatelliteEphemerisData);
exports.RtcmMessageGalileoINavSatelliteEphemerisData = RtcmMessageGalileoINavSatelliteEphemerisData;
let RtcmMessageIrnssSatelliteEphemerisData = class RtcmMessageIrnssSatelliteEphemerisData extends rtcm_1.RtcmMessage {
};
__decorate([
    decode_encode_1.DF(516),
    decode_encode_1.UInt(6),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "satelliteId", void 0);
__decorate([
    decode_encode_1.DF(517),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "weekNumber", void 0);
__decorate([
    decode_encode_1.DF(518),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "a_f0", void 0);
__decorate([
    decode_encode_1.DF(519),
    decode_encode_1.Int(16),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "a_f1", void 0);
__decorate([
    decode_encode_1.DF(520),
    decode_encode_1.Int(8),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "a_f2", void 0);
__decorate([
    decode_encode_1.DF(521),
    decode_encode_1.UInt(4),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "ura", void 0);
__decorate([
    decode_encode_1.DF(522),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "t_oc", void 0);
__decorate([
    decode_encode_1.DF(523),
    decode_encode_1.Int(8),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "t_gd", void 0);
__decorate([
    decode_encode_1.DF(524),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "deltaN", void 0);
__decorate([
    decode_encode_1.DF(525),
    decode_encode_1.UInt(8),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "iodec", void 0);
__decorate([
    decode_encode_1.DF(526),
    decode_encode_1.UInt(10),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "reservedAfterIodec", void 0);
__decorate([
    decode_encode_1.DF(527),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "l5Flag", void 0);
__decorate([
    decode_encode_1.DF(528),
    decode_encode_1.Bool,
    __metadata("design:type", Boolean)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "sFlag", void 0);
__decorate([
    decode_encode_1.DF(529),
    decode_encode_1.Int(15),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "c_uc", void 0);
__decorate([
    decode_encode_1.DF(530),
    decode_encode_1.Int(15),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "c_us", void 0);
__decorate([
    decode_encode_1.DF(531),
    decode_encode_1.Int(15),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "c_ic", void 0);
__decorate([
    decode_encode_1.DF(532),
    decode_encode_1.Int(15),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "c_is", void 0);
__decorate([
    decode_encode_1.DF(533),
    decode_encode_1.Int(15),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "c_rc", void 0);
__decorate([
    decode_encode_1.DF(534),
    decode_encode_1.Int(15),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "c_rs", void 0);
__decorate([
    decode_encode_1.DF(535),
    decode_encode_1.Int(14),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "idot", void 0);
__decorate([
    decode_encode_1.DF(536),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "m_0", void 0);
__decorate([
    decode_encode_1.DF(537),
    decode_encode_1.UInt(16),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "t_oe", void 0);
__decorate([
    decode_encode_1.DF(538),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "e", void 0);
__decorate([
    decode_encode_1.DF(539),
    decode_encode_1.UInt(32),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "sqrt_a", void 0);
__decorate([
    decode_encode_1.DF(540),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "omega_0", void 0);
__decorate([
    decode_encode_1.DF(541),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "w", void 0);
__decorate([
    decode_encode_1.DF(542),
    decode_encode_1.Int(22),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "omegaDot", void 0);
__decorate([
    decode_encode_1.DF(543),
    decode_encode_1.Int(32),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "i_0", void 0);
__decorate([
    decode_encode_1.DF(544),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "spareBitsAfterIdot", void 0);
__decorate([
    decode_encode_1.DF(545),
    decode_encode_1.UInt(2),
    __metadata("design:type", Number)
], RtcmMessageIrnssSatelliteEphemerisData.prototype, "spareBitsAfterI0", void 0);
RtcmMessageIrnssSatelliteEphemerisData = __decorate([
    rtcm_1.MessageType(rtcm_1.RtcmMessageType.IRNSS_SATELLITE_EPHEMERIS_DATA),
    rtcm_1.Since(rtcm_1.RtcmVersion.V3_3),
    rtcm_1.NavSystem(rtcm_1.RtcmNavSystem.IRNSS)
], RtcmMessageIrnssSatelliteEphemerisData);
exports.RtcmMessageIrnssSatelliteEphemerisData = RtcmMessageIrnssSatelliteEphemerisData;
