// @gnss/nmea/build/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NmeaResidualsMode = exports.NmeaNavStatusIndicator = exports.NmeaDataValidity = exports.NmeaOperationMode = exports.NmeaNavMode = exports.NmeaNavSystem = exports.NmeaDatum = exports.NmeaModeIndicator = exports.NmeaFixQuality = void 0;
var NmeaFixQuality;
(function (NmeaFixQuality) {
    NmeaFixQuality[NmeaFixQuality["INVALID"] = 0] = "INVALID";
    NmeaFixQuality[NmeaFixQuality["NORMAL"] = 1] = "NORMAL";
    NmeaFixQuality[NmeaFixQuality["DIFFERENTIAL"] = 2] = "DIFFERENTIAL";
    NmeaFixQuality[NmeaFixQuality["PPS"] = 3] = "PPS";
    NmeaFixQuality[NmeaFixQuality["RTK_INTEGER"] = 4] = "RTK_INTEGER";
    NmeaFixQuality[NmeaFixQuality["RTK_FLOAT"] = 5] = "RTK_FLOAT";
    NmeaFixQuality[NmeaFixQuality["DEAD_RECKONING"] = 6] = "DEAD_RECKONING";
    NmeaFixQuality[NmeaFixQuality["MANUAL"] = 7] = "MANUAL";
    NmeaFixQuality[NmeaFixQuality["SIMULATOR"] = 8] = "SIMULATOR";
})(NmeaFixQuality = exports.NmeaFixQuality || (exports.NmeaFixQuality = {}));
var NmeaModeIndicator;
(function (NmeaModeIndicator) {
    NmeaModeIndicator["AUTONOMOUS"] = "A";
    NmeaModeIndicator["DIFFERENTIAL"] = "D";
    NmeaModeIndicator["ESTIMATED"] = "E";
    NmeaModeIndicator["RTK_FLOAT"] = "F";
    NmeaModeIndicator["MANUAL"] = "M";
    NmeaModeIndicator["NO_FIX"] = "N";
    NmeaModeIndicator["PRECISE"] = "P";
    NmeaModeIndicator["RTK_FIXED"] = "R";
    NmeaModeIndicator["SIMULATOR"] = "S";
})(NmeaModeIndicator = exports.NmeaModeIndicator || (exports.NmeaModeIndicator = {}));
var NmeaDatum;
(function (NmeaDatum) {
    NmeaDatum["WGS84"] = "W84";
    NmeaDatum["WGS72"] = "W72";
    NmeaDatum["SGS85"] = "S85";
    NmeaDatum["PE90"] = "P90";
    NmeaDatum["USER_DEFINED"] = "999";
})(NmeaDatum = exports.NmeaDatum || (exports.NmeaDatum = {}));
var NmeaNavSystem;
(function (NmeaNavSystem) {
    NmeaNavSystem[NmeaNavSystem["GPS"] = 1] = "GPS";
    NmeaNavSystem[NmeaNavSystem["GLONASS"] = 2] = "GLONASS";
    NmeaNavSystem[NmeaNavSystem["GALILEO"] = 3] = "GALILEO";
    NmeaNavSystem[NmeaNavSystem["BEIDOU"] = 4] = "BEIDOU";
})(NmeaNavSystem = exports.NmeaNavSystem || (exports.NmeaNavSystem = {}));
var NmeaNavMode;
(function (NmeaNavMode) {
    NmeaNavMode[NmeaNavMode["UNAVAILABLE"] = 1] = "UNAVAILABLE";
    NmeaNavMode[NmeaNavMode["FIX_2D"] = 2] = "FIX_2D";
    NmeaNavMode[NmeaNavMode["FIX_3D"] = 3] = "FIX_3D";
})(NmeaNavMode = exports.NmeaNavMode || (exports.NmeaNavMode = {}));
var NmeaOperationMode;
(function (NmeaOperationMode) {
    NmeaOperationMode["MANUAL"] = "M";
    NmeaOperationMode["AUTOMATIC"] = "A";
})(NmeaOperationMode = exports.NmeaOperationMode || (exports.NmeaOperationMode = {}));
var NmeaDataValidity;
(function (NmeaDataValidity) {
    NmeaDataValidity["VALID"] = "A";
    NmeaDataValidity["INVALID"] = "V";
})(NmeaDataValidity = exports.NmeaDataValidity || (exports.NmeaDataValidity = {}));
var NmeaNavStatusIndicator;
(function (NmeaNavStatusIndicator) {
    NmeaNavStatusIndicator["SAFE"] = "S";
    NmeaNavStatusIndicator["CAUTION"] = "C";
    NmeaNavStatusIndicator["UNSAFE"] = "U";
    NmeaNavStatusIndicator["UNAVAILABLE"] = "V";
})(NmeaNavStatusIndicator = exports.NmeaNavStatusIndicator || (exports.NmeaNavStatusIndicator = {}));
var NmeaResidualsMode;
(function (NmeaResidualsMode) {
    NmeaResidualsMode[NmeaResidualsMode["PRE"] = 0] = "PRE";
    NmeaResidualsMode[NmeaResidualsMode["POST"] = 1] = "POST";
})(NmeaResidualsMode = exports.NmeaResidualsMode || (exports.NmeaResidualsMode = {}));
