// @gnss/nmea/build/mnemonics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NmeaSentenceType = void 0;
var NmeaSentenceType;
(function (NmeaSentenceType) {
    NmeaSentenceType["AAM"] = "AAM";
    NmeaSentenceType["ACK"] = "ACK";
    NmeaSentenceType["ADS"] = "ADS";
    NmeaSentenceType["AKD"] = "AKD";
    NmeaSentenceType["ALA"] = "ALA";
    NmeaSentenceType["ALM"] = "ALM";
    NmeaSentenceType["ALR"] = "ALR";
    NmeaSentenceType["APB"] = "APB";
    NmeaSentenceType["BEC"] = "BEC";
    NmeaSentenceType["BOD"] = "BOD";
    NmeaSentenceType["BWC"] = "BWC";
    NmeaSentenceType["BWR"] = "BWR";
    NmeaSentenceType["BWW"] = "BWW";
    NmeaSentenceType["CEK"] = "CEK";
    NmeaSentenceType["COP"] = "COP";
    NmeaSentenceType["CUR"] = "CUR";
    NmeaSentenceType["DBT"] = "DBT";
    NmeaSentenceType["DCN"] = "DCN";
    NmeaSentenceType["DCR"] = "DCR";
    NmeaSentenceType["DDC"] = "DDC";
    NmeaSentenceType["DOR"] = "DOR";
    NmeaSentenceType["DPT"] = "DPT";
    NmeaSentenceType["DSC"] = "DSC";
    NmeaSentenceType["DSE"] = "DSE";
    NmeaSentenceType["DSI"] = "DSI";
    NmeaSentenceType["DSR"] = "DSR";
    NmeaSentenceType["DTM"] = "DTM";
    NmeaSentenceType["ETL"] = "ETL";
    NmeaSentenceType["EVE"] = "EVE";
    NmeaSentenceType["FIR"] = "FIR";
    NmeaSentenceType["FSI"] = "FSI";
    NmeaSentenceType["GBS"] = "GBS";
    NmeaSentenceType["GEN"] = "GEN";
    NmeaSentenceType["GFA"] = "GFA";
    NmeaSentenceType["GGA"] = "GGA";
    NmeaSentenceType["GLC"] = "GLC";
    NmeaSentenceType["GLL"] = "GLL";
    NmeaSentenceType["GMP"] = "GMP";
    NmeaSentenceType["GNS"] = "GNS";
    NmeaSentenceType["GRS"] = "GRS";
    NmeaSentenceType["GSA"] = "GSA";
    NmeaSentenceType["GST"] = "GST";
    NmeaSentenceType["GSV"] = "GSV";
    NmeaSentenceType["HBT"] = "HBT";
    NmeaSentenceType["HDG"] = "HDG";
    NmeaSentenceType["HDT"] = "HDT";
    NmeaSentenceType["HMR"] = "HMR";
    NmeaSentenceType["HMS"] = "HMS";
    NmeaSentenceType["HSC"] = "HSC";
    NmeaSentenceType["HSS"] = "HSS";
    NmeaSentenceType["HTC"] = "HTC";
    NmeaSentenceType["HTD"] = "HTD";
    NmeaSentenceType["LCD"] = "LCD";
    NmeaSentenceType["MLA"] = "MLA";
    NmeaSentenceType["MSK"] = "MSK";
    NmeaSentenceType["MSS"] = "MSS";
    NmeaSentenceType["MTW"] = "MTW";
    NmeaSentenceType["MWD"] = "MWD";
    NmeaSentenceType["MWV"] = "MWV";
    NmeaSentenceType["NAK"] = "NAK";
    NmeaSentenceType["NRM"] = "NRM";
    NmeaSentenceType["NRX"] = "NRX";
    NmeaSentenceType["OSD"] = "OSD";
    NmeaSentenceType["POS"] = "POS";
    NmeaSentenceType["PRC"] = "PRC";
    NmeaSentenceType["RMA"] = "RMA";
    NmeaSentenceType["RMB"] = "RMB";
    NmeaSentenceType["RMC"] = "RMC";
    NmeaSentenceType["ROR"] = "ROR";
    NmeaSentenceType["ROT"] = "ROT";
    NmeaSentenceType["RPM"] = "RPM";
    NmeaSentenceType["RSA"] = "RSA";
    NmeaSentenceType["RSD"] = "RSD";
    NmeaSentenceType["RST"] = "RST";
    NmeaSentenceType["RTE"] = "RTE";
    NmeaSentenceType["SFI"] = "SFI";
    NmeaSentenceType["SID"] = "SID";
    NmeaSentenceType["STN"] = "STN";
    NmeaSentenceType["THS"] = "THS";
    NmeaSentenceType["TLB"] = "TLB";
    NmeaSentenceType["TLL"] = "TLL";
    NmeaSentenceType["TRC"] = "TRC";
    NmeaSentenceType["TRD"] = "TRD";
    NmeaSentenceType["TTM"] = "TTM";
    NmeaSentenceType["TUT"] = "TUT";
    NmeaSentenceType["TXT"] = "TXT";
    NmeaSentenceType["UID"] = "UID";
    NmeaSentenceType["VBW"] = "VBW";
    NmeaSentenceType["VDR"] = "VDR";
    NmeaSentenceType["VER"] = "VER";
    NmeaSentenceType["VHW"] = "VHW";
    NmeaSentenceType["VLW"] = "VLW";
    NmeaSentenceType["VPW"] = "VPW";
    NmeaSentenceType["VTG"] = "VTG";
    NmeaSentenceType["WAT"] = "WAT";
    NmeaSentenceType["WCV"] = "WCV";
    NmeaSentenceType["WNC"] = "WNC";
    NmeaSentenceType["WPL"] = "WPL";
    NmeaSentenceType["XDR"] = "XDR";
    NmeaSentenceType["XTE"] = "XTE";
    NmeaSentenceType["XTR"] = "XTR";
    NmeaSentenceType["ZDA"] = "ZDA";
    NmeaSentenceType["ZDL"] = "ZDL";
    NmeaSentenceType["ZFO"] = "ZFO";
    NmeaSentenceType["ZTG"] = "ZTG";
})(NmeaSentenceType = exports.NmeaSentenceType || (exports.NmeaSentenceType = {}));
