// @gnss/nmea/build/talkers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NmeaTalker = void 0;
var NmeaTalker;
(function (NmeaTalker) {
    NmeaTalker["AIS_BASE_STATION_INDEPENDENT"] = "AB";
    NmeaTalker["AIS_BASE_STATION_DEPENDENT"] = "AD";
    NmeaTalker["HEADING_TRACK_CONTROLLER_GENERAL"] = "AG";
    NmeaTalker["HEADING_TRACK_CONTROLLER_MAGNETIC"] = "AP";
    NmeaTalker["AIS_MOBILE_STATION_CLASS_AB"] = "AI";
    NmeaTalker["AIS_AIDS_NAVIGATION_STATION"] = "AN";
    NmeaTalker["AIS_RECEIVING_STATION"] = "AR";
    NmeaTalker["AIS_STATION_LIMITED"] = "AS";
    NmeaTalker["AIS_TRANSMITTING_STATION"] = "AT";
    NmeaTalker["AIS_SIMPLEX_REPEATER_STATION"] = "AX";
    NmeaTalker["BILGE_SYSTEMS"] = "BI";
    NmeaTalker["BRIDGE_NAVIGATIONAL_WATCH_ALARM_SYSTEM"] = "BN";
    NmeaTalker["CENTRAL_ALARM_MANAGEMENT"] = "CA";
    NmeaTalker["COMMUNICATIONS_DIGITAL_SELECTIVE_CALLING"] = "CD";
    NmeaTalker["COMMUNICATIONS_DATA_RECEIVER"] = "CR";
    NmeaTalker["COMMUNICATIONS_SATELLITE"] = "CS";
    NmeaTalker["COMMUNICATIONS_RADIO_TELEPHONE_MF_HF"] = "CT";
    NmeaTalker["COMMUNICATIONS_RADIO_TELEPHONE_VHF"] = "CV";
    NmeaTalker["COMMUNICATIONS_SCANNING_RECEIVER"] = "CX";
    NmeaTalker["DIRECTION_FINDER"] = "DF";
    NmeaTalker["DUPLEX_REPEATER_STATIONS"] = "DU";
    NmeaTalker["DYNAMIC_POSITION"] = "DP";
    NmeaTalker["ELECTRONIC_CHART_SYSTEM"] = "EC";
    NmeaTalker["ELECTRONIC_CHART_DISPLAY_INFORMATION_SYSTEM"] = "EI";
    NmeaTalker["EMERGENCY_POSITION_INDICATING_BEACON"] = "EP";
    NmeaTalker["ENGINE_ROOM_MONITORING_SYSTEMS"] = "ER";
    NmeaTalker["FIRE_DOOR_CONTROLLER_MONITORING_POINT"] = "FD";
    NmeaTalker["FIRE_EXTINGUISHER_SYSTEM"] = "FE";
    NmeaTalker["FIRE_DETECTION_POINT"] = "FR";
    NmeaTalker["FIRE_SPRINKLER_SYSTEM"] = "FS";
    NmeaTalker["NAV_SYSTEM_GALILEO"] = "GA";
    NmeaTalker["NAV_SYSTEM_BEIDOU"] = "GB";
    NmeaTalker["NAV_SYSTEM_NAVIC"] = "GI";
    NmeaTalker["NAV_SYSTEM_GLONASS"] = "GL";
    NmeaTalker["NAV_SYSTEM_GNSS"] = "GN";
    NmeaTalker["NAV_SYSTEM_GPS"] = "GP";
    NmeaTalker["NAV_SYSTEM_QZSS"] = "GQ";
    NmeaTalker["HEADING_SENSOR_COMPASS_MAGNETIC"] = "HC";
    NmeaTalker["HEADING_SENSOR_GYRO_NORTH_SEEKING"] = "HE";
    NmeaTalker["HEADING_SENSOR_FLUXGATE"] = "HF";
    NmeaTalker["HEADING_SENSOR_GYRO_NON_NORTH_SEEKING"] = "HN";
    NmeaTalker["HULL_DOOR_CONTROLLER_MONITORING_PANEL"] = "HD";
    NmeaTalker["HULL_STRESS_MONITORING"] = "HS";
    NmeaTalker["INTEGRATED_INSTRUMENTATION"] = "II";
    NmeaTalker["INTEGRATED_NAVIGATION"] = "IN";
    NmeaTalker["AUTOMATION_ALARM_MONITORING_SYSTEM"] = "JA";
    NmeaTalker["AUTOMATION_REEFER_MONITORING_SYSTEM"] = "JB";
    NmeaTalker["AUTOMATION_POWER_MANAGEMENT_SYSTEM"] = "JC";
    NmeaTalker["AUTOMATION_PROPULSION_CONTROL_SYSTEM"] = "JD";
    NmeaTalker["AUTOMATION_ENGINE_CONTROL_CONSOLE"] = "JE";
    NmeaTalker["AUTOMATION_PROPULSION_BOILER"] = "JF";
    NmeaTalker["AUTOMATION_AUXILIARY_BOILER"] = "JG";
    NmeaTalker["AUTOMATION_ELECTRONIC_GOVERNOR_SYSTEM"] = "JH";
    NmeaTalker["LORAN_C"] = "LC";
    NmeaTalker["MULTIPLEXER"] = "MX";
    NmeaTalker["NAVIGATION_LIGHT_CONTROLLER"] = "NL";
    NmeaTalker["PROPRIETARY_CODE"] = "P";
    NmeaTalker["RADAR"] = "RA";
    NmeaTalker["RECORD_BOOK"] = "RB";
    NmeaTalker["PROPULSION_MACHINERY_REMOTE_CONTROL"] = "RC";
    NmeaTalker["RUDDER_ANGLE_INDICATOR"] = "RI";
    NmeaTalker["PHYSICAL_SHORE_AIS_STATION"] = "SA";
    NmeaTalker["STEERING_CONTROL_SYSTEM"] = "SB";
    NmeaTalker["SOUNDER_DEPTH"] = "SD";
    NmeaTalker["TRACK_CONTROL_SYSTEM"] = "TC";
    NmeaTalker["TURN_RATE_INDICATOR"] = "TI";
    NmeaTalker["MICROPROCESSOR_CONTROLLER"] = "UP";
    NmeaTalker["USER_CONFIGURED_TALKER_1"] = "U1";
    NmeaTalker["USER_CONFIGURED_TALKER_2"] = "U2";
    NmeaTalker["USER_CONFIGURED_TALKER_3"] = "U3";
    NmeaTalker["USER_CONFIGURED_TALKER_4"] = "U4";
    NmeaTalker["USER_CONFIGURED_TALKER_5"] = "U5";
    NmeaTalker["USER_CONFIGURED_TALKER_6"] = "U6";
    NmeaTalker["USER_CONFIGURED_TALKER_7"] = "U7";
    NmeaTalker["USER_CONFIGURED_TALKER_8"] = "U8";
    NmeaTalker["USER_CONFIGURED_TALKER_9"] = "U9";
    NmeaTalker["VELOCITY_SENSOR_DOPPLER"] = "VD";
    NmeaTalker["VELOCITY_SENSOR_SPEED_LOG_WATER_MAGNETIC"] = "VM";
    NmeaTalker["VELOCITY_SENSOR_SPEED_LOG_WATER_MECHANICAL"] = "VW";
    NmeaTalker["VHF_DATA_EXCHANGE_SYSTEM_ASM"] = "VA";
    NmeaTalker["VHF_DATA_EXCHANGE_SYSTEM_SATELLITE"] = "VS";
    NmeaTalker["VHF_DATA_EXCHANGE_SYSTEM_TERRESTRIAL"] = "VT";
    NmeaTalker["VOYAGE_DATA_RECORDER"] = "VR";
    NmeaTalker["WATERTIGHT_DOOR_CONTROLLER_MONITORING_PANEL"] = "WD";
    NmeaTalker["WEATHER_INSTRUMENTS"] = "WI";
    NmeaTalker["WATER_LEVEL_DETECTION_SYSTEM"] = "WL";
    NmeaTalker["TRANSDUCER"] = "YX";
    NmeaTalker["TIMEKEEPER_ATOMIC_CLOCK"] = "ZA";
    NmeaTalker["TIMEKEEPER_CHRONOMETER"] = "ZC";
    NmeaTalker["TIMEKEEPER_QUARTZ"] = "ZQ";
    NmeaTalker["TIMEKEEPER_RADIO_UPDATE"] = "ZV";
})(NmeaTalker = exports.NmeaTalker || (exports.NmeaTalker = {}));
